{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Structure, Systems, Dataset and Loader Tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Topics:\n",
    "- Interacting with `Structure` class\n",
    "- Interacting `PlinderSystem`\n",
    "- Interacting `PlinderDataset`\n",
    "- Loader\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Setup\n",
    "\n",
    "### Installation\n",
    "\n",
    "`plinder` is available on *PyPI*.\n",
    "\n",
    "```\n",
    "pip install plinder\n",
    "```\n",
    "\n",
    "### Environment variable configuration\n",
    ":::{note}\n",
    "We need to set environment variables to point to the release and iteration of choice.\n",
    "For the sake of demonstration, this will be set to point to a smaller tutorial example\n",
    "dataset, which are `PLINDER_RELEASE=2024-06` and `PLINDER_ITERATION=tutorial`.\n",
    ":::\n",
    ":::{note}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting the configuration\n",
    "\n",
    "At first we get the configuration to check that all parameters are correctly set. \n",
    "In the snippet below, we will check, if the local and remote *PLINDER* paths point to\n",
    "the expected location."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "local cache directory: /Users/vladas/.local/share/plinder/2024-06/v2\n",
      "remote data directory: gs://plinder/2024-06/v2\n"
     ]
    }
   ],
   "source": [
    "import plinder.core.utils.config\n",
    "\n",
    "cfg = plinder.core.get_config()\n",
    "print(f\"local cache directory: {cfg.data.plinder_dir}\")\n",
    "print(f\"remote data directory: {cfg.data.plinder_remote}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data ecosystem overview\n",
    "This tutorial assumes user have downloaded _PLINDER_ dataset before now. While the examples will run without users doing anything, we encourage users to download the data for performance sake. _PLINDER_ data hierarchy is shown below. We have organized this tutorial to follow this same hierarchy from ground up\n",
    "![image](../static/asset/data/plinder_data_hierarchy.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Structure files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After download all files will be store locally at `~/.local/share/plinder/${PLINDER_RELEASE}/${PLINDER_ITERATION}/systems`. The current default is `PLINDER_RELEASE=2024-06` and `PLINDER_ITERATION-v2`\n",
    "\n",
    "There, we have sub-folders that corresponds to each systems. In each sub-folder, we have:\n",
    "- Receptor PDB: receptor.cif and receptor.pdb\n",
    "- Ligand SDF's: `<biounit_instance_id>.<chain_id>.sdf`. For complexes with more than one ligands, all the sdfs are saved separately.\n",
    "- Sequence fasta: sequence.fasta\n",
    "For more information on the file organization, see \"<link-to-dataset-tutorial>\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Structure Python Abstraction\n",
    "To make interacting with our data seamless, {class} class Structure, a pydantic data class that:\n",
    "- Loads all the structure files + smiles \n",
    "- Gets coordinates\n",
    "- Featurizes residues and atoms of associated protein and ligand molecules\n",
    "- Masks molecules to account for resolved vs unresolved part\n",
    "\n",
    "To interact with the example, do the following:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the structure for a given system_id\n",
    "For this purpose we will use `\"1avd__1__1.A__1.C\"` as our example system id."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-23 12:30:34,946 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 12:30:34,946 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 12:30:35,399 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 12:30:35,401 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 12:30:35,402 | plinder.core.index.utils:148 | INFO : loading entries from 1 zips\n",
      "2024-09-23 12:30:35,410 | plinder.core.index.utils:163 | INFO : loaded 1 entries\n",
      "2024-09-23 12:30:35,411 | plinder.core.index.utils.load_entries:24 | INFO : runtime succeeded: 0.46s\n"
     ]
    }
   ],
   "source": [
    "from plinder.core.structure.structure import Structure\n",
    "from plinder.core import PlinderSystem\n",
    "from pathlib import Path\n",
    "\n",
    "from biotite.sequence.io.fasta import FastaFile\n",
    "\n",
    "input_smiles = \"CC(=O)N[C@@H]1[C@H]([C@@H]([C@H](O[C@H]1O)CO)O)O\" # Need to account for unresolved part of the ligand\n",
    "input_sdf = Path(cfg.data.plinder_dir)/\"systems/1avd__1__1.A__1.C/ligand_files/1.C.sdf\"\n",
    "system_id = \"1avd__1__1.A__1.C\"\n",
    "protein_structure_path = Path(cfg.data.plinder_dir)/\"systems/1avd__1__1.A__1.C/receptor.cif\"\n",
    "input_sequence_path = Path(cfg.data.plinder_dir)/\"systems/1avd__1__1.A__1.C/sequences.fasta\"\n",
    "list_ligand_sdf_and_input_smiles = [\n",
    "    (input_sdf,input_smiles)]\n",
    "\n",
    "input_sequences = {k: v for k, v in FastaFile.read_iter(input_sequence_path)}\n",
    "\n",
    "# Load holo structure\n",
    "holo_struc = PlinderSystem(system_id=system_id).holo_structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': FieldInfo(annotation=str, required=True),\n",
       " 'protein_path': FieldInfo(annotation=Path, required=True),\n",
       " 'protein_sequence': FieldInfo(annotation=Union[dict[str, str], NoneType], required=False, default=None),\n",
       " 'ligand_sdfs': FieldInfo(annotation=Union[dict[str, str], NoneType], required=False, default=None),\n",
       " 'ligand_smiles': FieldInfo(annotation=Union[dict[str, str], NoneType], required=False, default=None),\n",
       " 'protein_atom_array': FieldInfo(annotation=Union[AtomArray, NoneType], required=False, default=None),\n",
       " 'ligand_mols': FieldInfo(annotation=Union[dict[str, tuple[Mol, Mol, Mol, tuple[ndarray[Any, dtype[+_ScalarType_co]], ndarray[Any, dtype[+_ScalarType_co]]]]], NoneType], required=False, default=None),\n",
       " 'add_ligand_hydrogens': FieldInfo(annotation=bool, required=False, default=False),\n",
       " 'structure_type': FieldInfo(annotation=str, required=False, default='holo')}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show fields\n",
    "holo_struc.model_fields"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List fields\n",
    "- We list all fields and their `FieldInfo` to show which ones are required. `id`, `protein_path` and `protein_sequence` are required. Everything else is optionally. Particularly worth mentioning is the decision to make `list_ligand_sdf_and_input_smiles` optional; this is because ligand will not be availbale in apo and predicted structures.\n",
    "- Out of these field `ligand_mols` and `protein_atom_array` is computed within the object if set to default. \n",
    "- `ligand_mols` returns a chain-mapped dictionary of of the form:\n",
    "    ```python\n",
    "    {\n",
    "        \"<instance_id>.<chain_id>\": (\n",
    "            RDKit 2D mol from template SMILES of type `Chem.Mol`,\n",
    "            RDKit mol from template SMILES with random 3D conformer of type `Chem.Mol`,\n",
    "            RDKit mol of solved (holo) ligand structure of type `Chem.Mol`,\n",
    "            paired stacked arrays (template vs holo) mapping atom order by index of type `tuple[NDArray.int_, NDArray.int_]`\n",
    "        )\n",
    "\n",
    "    }\n",
    "    ```\n",
    "- While `protein_atom_array` returns [biotite AtomArray](https://www.biotite-python.org/latest/apidoc/biotite.structure.AtomArray.html) of the receptor protein structure.\n",
    "- `add_ligand_hydrogens` specifies whether to adds hydrogens to ligand\n",
    "- `structure_type`: could be `\"holo\"`, `\"apo\"` or `\"pred\"`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1.C': (<rdkit.Chem.rdchem.Mol at 0x1ca05f5a0>,\n",
       "  <rdkit.Chem.rdchem.Mol at 0x1c9de46d0>,\n",
       "  <rdkit.Chem.rdchem.Mol at 0x1c9de45f0>,\n",
       "  (array([[13,  4,  5,  7,  9, 10,  1,  0,  3,  6,  8, 12, 11,  2]]),\n",
       "   array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13]])))}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Inspect ligand_mols\n",
    "holo_struc.ligand_mols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1.C': 'CC(=O)N[C@@H]1[C@@H](O)[C@H](O)[C@@H](CO)O[C@H]1O'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.ligand_smiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inspect protein_atom_araay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Atom(np.array([31.221, 22.957, 43.101], dtype=float32), chain_id=\"1.A\", res_id=3, ins_code=\"\", res_name=\"LYS\", hetero=False, atom_name=\"N\", element=\"N\", b_factor=0.0)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_atom_array[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List structure protein properties\n",
    "Show protein related properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "protein_backbone_mask\n",
      "protein_calpha_coords\n",
      "protein_calpha_mask\n",
      "protein_chain_ordered\n",
      "protein_chains\n",
      "protein_coords\n",
      "protein_n_atoms\n",
      "protein_sequence_from_structure\n",
      "protein_structure_b_factor\n",
      "protein_structure_tokenized_sequence\n",
      "protein_unique_atom_names\n",
      "protein_unique_residue_ids\n",
      "protein_unique_residue_names\n"
     ]
    }
   ],
   "source": [
    "for property in holo_struc.get_properties():\n",
    "    if \"protein\" in property:\n",
    "        print(property)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Protein backbone mask\n",
    "This is a boolean mask that can be used to select backbone atoms from biotite `AtomArray`. The indices of `True` corresponds to backbone indices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False,  True,  True,  True,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "        True,  True,  True, False,  True,  True,  True, False, False,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False,  True,  True,  True, False,  True,\n",
       "        True,  True, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "        True,  True,  True, False,  True,  True,  True, False, False,\n",
       "        True,  True,  True, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "        True,  True,  True, False,  True,  True,  True, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False,  True,  True,  True, False, False,  True,  True,  True,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False,  True,  True,  True, False, False,  True,  True,\n",
       "        True, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False, False, False, False,  True,  True,  True, False,  True,\n",
       "        True,  True, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False,  True,\n",
       "        True,  True, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False, False, False, False, False,  True,\n",
       "        True,  True, False,  True,  True,  True, False, False, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False, False, False, False,  True,  True,\n",
       "        True, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False,  True,\n",
       "        True,  True, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False,  True,  True,  True, False,  True,\n",
       "        True,  True, False, False, False, False, False, False,  True,\n",
       "        True,  True, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False, False,  True,  True,  True,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False, False,  True,  True,  True,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False,  True,\n",
       "        True,  True, False, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False,  True,  True,  True, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False,  True,\n",
       "        True,  True, False, False, False, False, False,  True,  True,\n",
       "        True, False, False, False, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False,  True,  True,  True,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False,  True,\n",
       "        True,  True, False, False, False, False, False,  True,  True,\n",
       "        True, False,  True,  True,  True, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False, False, False, False, False,  True,  True,  True, False,\n",
       "       False,  True,  True,  True, False, False, False, False,  True,\n",
       "        True,  True, False, False, False, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False,  True,\n",
       "        True,  True, False,  True,  True,  True, False, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "        True,  True,  True, False, False, False, False, False, False,\n",
       "       False, False,  True,  True,  True, False, False, False, False,\n",
       "       False,  True,  True,  True, False, False, False, False, False,\n",
       "       False, False, False,  True,  True,  True, False, False, False,\n",
       "       False])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_backbone_mask"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Protein Calpha mask\n",
    "This shows the mask of calpha atoms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False,  True, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False,  True, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "        True, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False,  True, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False,  True, False, False, False,  True, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "        True, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "        True, False, False, False,  True, False, False, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "        True, False, False, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False, False,  True, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False,  True, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False,  True, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False, False, False, False, False,  True, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "        True, False, False, False, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "        True, False, False, False,  True, False, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False,  True, False, False, False, False, False, False, False,\n",
       "       False, False, False,  True, False, False, False, False, False,\n",
       "       False, False,  True, False, False, False, False, False, False,\n",
       "       False, False, False, False,  True, False, False, False, False,\n",
       "       False])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_calpha_mask"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein chain ordered\n",
    "This gives a list of protein chains ordered by how they are in the structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1.A']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_chain_ordered"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein chain ordered\n",
    "This gives a list of protein chains ordered by how they are in the structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein chains for all atoms\n",
    "The list of chain IDs in the structure. Order of how they appear not kept."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1.A']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_chains"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein coordinates\n",
    "This property gets the 3D positions of each of the atoms in protein molecules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array([[31.221, 22.957, 43.101],\n",
       "        [31.828, 24.118, 42.476],\n",
       "        [31.979, 23.854, 41.021],\n",
       "        ...,\n",
       "        [34.341, 35.018, 24.674],\n",
       "        [35.484, 35.831, 24.497],\n",
       "        [33.105, 35.742, 24.15 ]], dtype=float32)]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_coords"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get number of atoms of protein molecule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "964"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_n_atoms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein structure atom names\n",
    "Returns all atoms names the same way they appear in the structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['C',\n",
       " 'CA',\n",
       " 'CB',\n",
       " 'CD',\n",
       " 'CD1',\n",
       " 'CD2',\n",
       " 'CE',\n",
       " 'CE1',\n",
       " 'CE2',\n",
       " 'CE3',\n",
       " 'CG',\n",
       " 'CG1',\n",
       " 'CG2',\n",
       " 'CH2',\n",
       " 'CZ',\n",
       " 'CZ2',\n",
       " 'CZ3',\n",
       " 'N',\n",
       " 'ND1',\n",
       " 'ND2',\n",
       " 'NE',\n",
       " 'NE1',\n",
       " 'NE2',\n",
       " 'NH1',\n",
       " 'NH2',\n",
       " 'NZ',\n",
       " 'O',\n",
       " 'OD1',\n",
       " 'OD2',\n",
       " 'OE1',\n",
       " 'OE2',\n",
       " 'OG',\n",
       " 'OG1',\n",
       " 'OH',\n",
       " 'SD',\n",
       " 'SG']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_unique_atom_names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein b-factors\n",
    "Get protein atom occupancies. If not available in a structure, it's set to zero."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0,\n",
       " 0.0]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_structure_b_factor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein residue names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ALA',\n",
       " 'ARG',\n",
       " 'ASN',\n",
       " 'ASP',\n",
       " 'CYS',\n",
       " 'GLN',\n",
       " 'GLU',\n",
       " 'GLY',\n",
       " 'HIS',\n",
       " 'ILE',\n",
       " 'LEU',\n",
       " 'LYS',\n",
       " 'MET',\n",
       " 'PHE',\n",
       " 'PRO',\n",
       " 'SER',\n",
       " 'THR',\n",
       " 'TRP',\n",
       " 'TYR',\n",
       " 'VAL']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_unique_residue_names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get protein residues number\n",
    "Residue number as they appear in structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[3,\n",
       " 4,\n",
       " 5,\n",
       " 6,\n",
       " 7,\n",
       " 8,\n",
       " 9,\n",
       " 10,\n",
       " 11,\n",
       " 12,\n",
       " 13,\n",
       " 14,\n",
       " 15,\n",
       " 16,\n",
       " 17,\n",
       " 18,\n",
       " 19,\n",
       " 20,\n",
       " 21,\n",
       " 22,\n",
       " 23,\n",
       " 24,\n",
       " 25,\n",
       " 26,\n",
       " 27,\n",
       " 28,\n",
       " 29,\n",
       " 30,\n",
       " 31,\n",
       " 32,\n",
       " 33,\n",
       " 34,\n",
       " 35,\n",
       " 36,\n",
       " 37,\n",
       " 38,\n",
       " 39,\n",
       " 40,\n",
       " 41,\n",
       " 42,\n",
       " 43,\n",
       " 44,\n",
       " 45,\n",
       " 46,\n",
       " 47,\n",
       " 48,\n",
       " 49,\n",
       " 50,\n",
       " 51,\n",
       " 52,\n",
       " 53,\n",
       " 54,\n",
       " 55,\n",
       " 56,\n",
       " 57,\n",
       " 58,\n",
       " 59,\n",
       " 60,\n",
       " 61,\n",
       " 62,\n",
       " 63,\n",
       " 64,\n",
       " 65,\n",
       " 66,\n",
       " 67,\n",
       " 68,\n",
       " 69,\n",
       " 70,\n",
       " 71,\n",
       " 72,\n",
       " 73,\n",
       " 74,\n",
       " 75,\n",
       " 76,\n",
       " 77,\n",
       " 78,\n",
       " 79,\n",
       " 80,\n",
       " 81,\n",
       " 82,\n",
       " 83,\n",
       " 84,\n",
       " 85,\n",
       " 86,\n",
       " 87,\n",
       " 88,\n",
       " 89,\n",
       " 90,\n",
       " 91,\n",
       " 92,\n",
       " 93,\n",
       " 94,\n",
       " 95,\n",
       " 96,\n",
       " 97,\n",
       " 98,\n",
       " 99,\n",
       " 100,\n",
       " 101,\n",
       " 102,\n",
       " 103,\n",
       " 104,\n",
       " 105,\n",
       " 106,\n",
       " 107,\n",
       " 108,\n",
       " 109,\n",
       " 110,\n",
       " 111,\n",
       " 112,\n",
       " 113,\n",
       " 114,\n",
       " 115,\n",
       " 116,\n",
       " 117,\n",
       " 118,\n",
       " 119,\n",
       " 120,\n",
       " 121,\n",
       " 122,\n",
       " 123,\n",
       " 124,\n",
       " 125]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_unique_residue_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get sequence from protein structure\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'KCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRT'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_sequence_from_structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE'}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get tokenized sequence\n",
    "Get tensor of sequence converted to integer-based amino acid token"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([11,  4, 15, 10, 16,  7, 11, 17, 16,  2,  3, 10,  7, 15,  2, 12, 16,  9,\n",
       "         7,  0, 19,  2, 15,  1,  7,  6, 13, 16,  7, 16, 18, 16, 16,  0, 19, 16,\n",
       "         0, 16, 15,  2,  6,  9, 11,  6, 15, 14, 10,  8,  7, 16,  6,  2, 16,  9,\n",
       "         2, 11,  1, 16,  5, 14, 16, 13,  7, 13, 16, 19,  2, 17, 11, 13, 15,  6,\n",
       "        15, 16, 16, 19, 13, 16,  7,  5,  4, 13,  9,  3,  1,  2,  7, 11,  6, 19,\n",
       "        10, 11, 16, 12, 17, 10, 10,  1, 15, 15, 19,  2,  3,  9,  7,  3,  3, 17,\n",
       "        11,  0, 16,  1, 19,  7,  9,  2,  9, 13, 16,  1, 10,  1, 16])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_structure_tokenized_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Inspect holo sequences\n",
    "Returns a chain-mapped dictionary of sequences from seqres\n",
    "```python\n",
    "{\n",
    "    \"<instance_id>.<chain_id>\": sequence of type `str`\n",
    "\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE'}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.protein_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Structure(\n",
       "    (\n",
       "        'id',\n",
       "        '1avd__1__1.A__1.C',\n",
       "    ),\n",
       "    (\n",
       "        'protein_path',\n",
       "        /Users/vladas/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/receptor.cif,\n",
       "    ),\n",
       "    (\n",
       "        'protein_sequence',\n",
       "        {\n",
       "            '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'ligand_sdfs',\n",
       "        {\n",
       "            '1.C': '/Users/vladas/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/ligand_files/1.C.sdf',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'ligand_smiles',\n",
       "        {\n",
       "            '1.C': 'CC(=O)N[C@@H]1[C@@H](O)[C@H](O)[C@@H](CO)O[C@H]1O',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'protein_atom_array',\n",
       "        <class 'biotite.structure.AtomArray'> with shape (123,),\n",
       "    ),\n",
       "    (\n",
       "        'ligand_mols',\n",
       "        {\n",
       "            '1.C': (\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x1d066bae0>,\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x1ca05fca0>,\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x1ca05e880>,\n",
       "                (\n",
       "                    <class 'numpy.ndarray'> with shape (1, 14),\n",
       "                    <class 'numpy.ndarray'> with shape (1, 14),\n",
       "                ),\n",
       "            ),\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'add_ligand_hydrogens',\n",
       "        False,\n",
       "    ),\n",
       "    (\n",
       "        'structure_type',\n",
       "        'holo',\n",
       "    ),\n",
       ")"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.filter(\n",
    "        property=\"atom_name\",\n",
    "        mask=\"CA\",\n",
    "\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List ligand properties\n",
    "Show ligand related properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input_ligand_conformers\n",
      "input_ligand_conformers_coords\n",
      "input_ligand_templates\n",
      "ligand_chain_ordered\n",
      "ligand_template2resolved_atom_order_stacks\n",
      "resolved_ligand_mols\n",
      "resolved_ligand_mols_coords\n"
     ]
    }
   ],
   "source": [
    "for property in holo_struc.get_properties():\n",
    "    if \"ligand\" in property:\n",
    "        print(property)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ":::{todo}\n",
    "\n",
    "The input ligands are provided using dictionaries.\n",
    "These dictionaries contain information for each ligand:\n",
    "- input_ligand_templates: 2D RDKit mols generated from RDKit canonical SMILES (taken from annotation table)\n",
    "- input_ligand_conformers: 3D (random) conformers generated for each input mol\n",
    "- input_ligand_conformers_coords: positional coordintates for 3D conformers\n",
    "- resolved_ligand_mols: RDKit mols of solved (holo) ligand structures\n",
    "- resolved_ligand_mols_coords: positional coordintates for holo ligand structures\n",
    "- ligand_template2resolved_atom_order_stacks: paired stacked arrays (template vs holo) mapping atom order by index\n",
    "- ligand_chain_ordered: ordered list of all ligands by their keys\n",
    "\n",
    ":::"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ligand atom id mapping mapping\n",
    "\n",
    "Unlike the protein sequence - there is no canonical order to ligand atoms in the molecule.\n",
    "It can be further complicated by automorphisms present in the structure due to symmetry, i.e. there is more than one match that is possible between the structures.\n",
    "\n",
    "This is important when calculating ligand structure loss, as the most optimal atom order can change between the different inference results. Typically, it is accepted to take the atom ordering resulting in the best objective score and use that for the loss calculation.\n",
    "\n",
    "Occasionally futher ambiguity arises to to part of the ligand structure being unresolved in the holo structure - this can lead to multiple available matches. We use RascalMCES algorithm from RDKit to provide all the possible matches between the atom order in the input structure (from SMILES) to the resolved holo structure.\n",
    "\n",
    "This is provided as stacks of atom order arrays that reorder the template and holo indices to provide matches. Each stack is a unique order transformation and should be iterated."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ligand conformer to input smiles mapping\n",
    "\n",
    "Each RDKit ligand mol that is generated from SMILES matches the atom order that is in the starting SMILES.\n",
    "This order is retained when 3D conformer is generated.\n",
    "\n",
    "! NOTE: While we ensure that each PLINDER ligand can be loaded and sanitized into RDKit 2D molecule, some ligands may struggle to generate sanitizeable 3D conformers. We nonetheless are trying to provide each ligand with as sensible starting structure as possible and make it accessible via coordinate arrays."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sample_system_2 = PlinderSystem(\n",
    "#     system_id=\"102m__1__1.A__1.C\",\n",
    "# )\n",
    "# sample_system_2.holo_structure.ligand_mols['1.C'][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1.C': (array([[13,  4,  5,  7,  9, 10,  1,  0,  3,  6,  8, 12, 11,  2]]),\n",
       "  array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13]]))}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.ligand_template2resolved_atom_order_stacks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below we use RDKit functionality to draw indixes for `'1.C'` ligand 2D, conformer and holo structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdkit.Chem.Draw import IPythonConsole\n",
    "IPythonConsole.drawOptions.addAtomIndices = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcIAAACWCAIAAADCEh9HAAAABmJLR0QA/wD/AP+gvaeTAAAfWUlEQVR4nO3deVxU9d4H8A8MOwIDsrmBQiyKiIkryjIwMKi4i1ezHstI63Zb1HrsZnWtbHsKbl6t+5Sa+2OuFKgJzoxCCi6YEGKA7CLLgKgMiDAw5/njjCxeRYSZOdPwfb/445wfh/P7zgv98ju/81sMGIYBIYSQ3jLkOgBCCPlzozRKCCF9QmmUEEL6hNIoIYT0CaVRQgjpE0qjhBDSJ5RGST+SnZ0dGxubm5vLdSBEr1AaJf1IbGzsa6+9FhcXx3UgRK9QGiX9yNChQ1955RUvLy+uAyF6hdIo6UdkMtm2bdvooZ6olwFNBiX9x759+0pLS729vefOnct1LER/UBolhJA+oYd6QgjpE0qjpB9QKJCVhcuX0dTUUVhWhubmjtObN1FXp/3QiB6gNEr03eHDGDYMy5fjlVcwbBg2bVKVT52KS5c6LvvgA/zP/3ASIPmzM+I6AEI0KS8Pzz2HhAQIhQCQlYXgYIwcqTolRB2oNUr02oEDEIk6kqafH2JisGcPpzERfUOtUaLXCgrg49OlZPRobNmiOl6/Hvb2quOMDMyfr9XYiL6gNEr0Go+HtrYuJW1tMLr/z376dIwcqTqWy7UaGNEjlEaJXvP0RGpql5LMTHh6qo4nTUJAgOo4MVGrgRE9Qn2jRK8tXozUVCQkqE4vXMDOnVi+nNOYiL6h1ijRa8OH49AhrFyJd96BqSkqKrBpE6ZM4TosoldoMijpjatXr+7evdvV1fXll1/mOpYeUCpRWIjWVnh4dHSM3rmDAQPA46lO2ZH55ubcREj+zKg1Snrj+++/j4qKcnV15TqQx3nnHZSUQCBAaGjH2ySWjU2XU0qgpLeob5T0hkwms7Gx+fjjj5s6T6/UQX/7G6ZPR1oawsLw5ptcR0P0E7VGSW/4+fmVlpYaGhoaGxtzHUu3hg7FsmVYtgwAdDzjkz8t6hslvcEwzKVLl9zd3W1tbbmO5dEUCty7BysrruMgeo7SKNFfly4hJAS+vhAI8OKLcHPjOiCin+ihnugvf3/U1CA9HadO4e5drqMheotao/1RUVGRra1tj57H9+yBpSXmzVOdXrmCgwfx4YcAwDA4fBiJibh7F25u+Otfofsv7gnRAHpT3+9IJJL09PSXXnqppaXl8VefPYuLFztOr1/H4cOq49Wr8c47CAtDTAwUCowbh5wcjUTca8nJyM/nOgii/+ihvt8JCwsDcO7cOaVS2fu75OXh22+Rm4sRIwBAJIJCgXXr8NNPagpTHZKScOgQGAb5+TAz4zoaoreoNdofHT58eNq0aWY9zCx37qCkRPUlk6kKL17EuHGqHMpatAgXLqg/1r6IjUVpKc6coRxKNIpao/3Ozp079+/fHxwcXFtba9++2mY3Dh9GerrquL5elZIqK+Hg0OUyJydUV6OtrWN6pY5wceE6AqLnKI32O8uWLVvGDkfvoeXL8emnquNffsHbbwPAkCGoru5yWWUlnJ11J4empKQkJydHRkYGBgZyHQvRc/RQT3pl4kRcvoyioo6Sgwd1auUkZ2dnHo8nkUi4DoToP2qNkl556im8/joiIrB+PQYNwvHj2L8fZ85wHVYHLy+vjz76iOsoSL9ArdH+obkZ69bB2xsDB2LiROza1dMfDArCpEkdpy4u+MtfVMdffonPP0dyMjZuhJERLl2Cl5fqWzIZAgJoDjvpJ2j4ff+wdCmuX8f//i/c3JCaiuefx0cfISZGI3XduoXwcFy5guHD8ccfMDDQSC3damlpGT169CeffBIdHa392kl/Q2m0HygpgZcXCgowbJiqZO9efPABCgvVX9ft24iIUI3YNzfHCy/gm2/UX0sPFBYWmpmZDRkyhJPaSb9CfaP9QE4Ohg/vyKEAgoNRXIzGRlhaqrmuJUs6Zj0pFNixA2PGYOVKNdfyH+Li4hQKxdChQ5cuXcqWuLu7a7pSQljUN9oP1Nc/mC6trMAwqK9Xf11790IgALsIaWsreDysX6+FV0+rV69+6aWXysvL2dOqqqrm5mZNV0oIi9JoPzBiBMrK0HnqZ1ERLCzg7Kz+uuzsIJHgtdfArnvS0IDGRsTEoKpK/XV1tWXLlvbxsFu2bKmpqdF0jYSwKI32A08/DWNj7NnTUbJxI2bM0NTLHwMDxMbin/+EiwvYnve8PCxYAE02D5ubm+VyufP9Pwzvv//+0KFDNVcdIZ3RK6b+4eRJREdjzhy4uyMlBeXlkEig6USTnY2lS5GdDWtrNDRg3jwcOqShqn7//XdTU1Ov9hFXhGgRtUb13b17OH8eoaHIy0NICMzM8OqryMrSeA4F4OuLM2cwfToUCjCM8vjxhCeahPokxowZY2xs/P3333t7e5uamo4cObL6gbmqhGgMtUb1nViM8HAEBSElBQDy8/HUUzDU4p9PpRIxMcoff/yVx/uLmVncxo3PPPOMuu4tl8ulUmlSUlJSUlJR55mpgI+PT3p6uhVtxEQ0jwY86TupFAAmTwaA+nr4+MDBAaWl0NqOnoaG+OGHjCFDlu/aVV1Wtm7dOh8fHz8/v77cMicn5+jRo2Kx+Ndff21/Iz9w4MDQ0NCJEyfy+fx169bl5OQIBILk5GQ7Ozt1fAxCHo0h+m3SJAZgfvmFYRgmMZEBmKlTOQnk6tWrTz/9NICRI0fKZLIn/XGZTHbgwIEVK1Z0HlHP4/H8/f3Xrl178uRJhULRfnFpaSk7bnTcuHG1tbVq/RyEPIjSqF6rr2eMjBgjI6a+nmEYZvVqBmDef5+7cOpFIpGpqenUqVObm5sfe31ra2tGRsY//vEPf39/w04dEc7Ozs8999yBAwfq6uoe9bNlZWVPPfUUgLFjx9bU1Kj1cxDSBaVRvZaQwADMtGmq07FjGYCRSjmMSKlUvvXWW7a2tgsXLnzUNYWFhd999110dLSNjU176jQ3NxcKhZ9//nlGRoZSqexJXZWVlaNGjQLg5+fXi/YvIT1EaVSvrVrFAMwHHzAMw9TWMoaGjJkZc/cu12Exe/bscXBweO+999pLGhsbT548uXbtWn9//86dTm5ubitWrDhw4IBcLu9FRVVVVaNHjwbg7e1dUVGhvk/wZDHcvn2bYZjr169nZ2dzEgPRKHrFpNfY90sCAQCcOgWlElOnwtyc26AALF261N/fXygUWlpapqSkyOXyixcvtu9UamdnJxQKIyIiRCJRH0fROzk5SSQSoVCYnZ0tEAikUungwYPV8Ql6qqSkJDY2NjIy0tfXNzY21tPTs7CwcM6cOdqMgWgaDXjSXzdvwtERJia4dUs1XPTbb7FhA9at4zCoFStWuLu7R0ZG+vn5FRQUjBo1SqFQAODxeGPHjhUKhUKhMDg42FitAwlqamqEQuHvv//u6ekpkUi0PMHp9OnTjY2NcrncxcVl8uTJq1at2rhxozYDIJpGrVH91d78ZDehY1umoaHcBtXY2DhgwAD2bfuhQ4cUCoWRkdH69ev/+te/2rLT8DXAwcHh9OnTIpHo4sWLgYGBUql0ROc9TbXC0tJSLpc3NjZaqn1VLcI1msWkv06dAu4/0VdWIjcXVlYYP57boH744YclS5Z89tlnAG7evAng+eefX7duneZyKMvW1jY5OXnSpEklJSUCgeCBsfqaU1dXFx8ff/z48cmTJyclJX344YfLly/XTtVEa6g1qr86Nz/Znd2CgrQ36v5hlErlV199ZWhoyA4glUqlABYvXqyd2vl8flJSUmRk5Llz5wQCgUQiYUdEac7t27e3bNlSWVm5detWa2vruLg4jVZHOMP1Oy6iEbcrKpo8PRlra4YdlL58OQMwX33FdVyMQqGor69nGObmzZuGhoampqZ3tTtyoKGhISQkBMDQoUPz8/M1UUVzc/PevXujoqLa0/TEiRPZl/VEL9FDvX46KpWa5+e/EB4OIyMATXl5wP0HfE4ZGRmx89xPnz6tVCoDAgLMtTtywNLS8ujRo6GhoeXl5YGBgTk5OWq8eXp6+jPPPDN+/Pjly5cfPXq0oKCALb9w4UJwcDDbiUH0D6VR/cQ+L4+eMgVAcXGxxdmzolGjMHYs13F1OHXqFAABF5nd0tIyMTFRKBRWV1eHhYVduXKl83dv3bq1cePGc+fO9fyGpaWlb7311rhx40Qi0b59+7Kzs/9z7f2srCyBQFBbW6uGD0B0DdfNYaIRw4cPB/Dbb78xDLNlyxYACxYs4DqoLtj5RWfOnOEqgLt374aHhwNwdHTMyspqL1+9enVNTU1MTExTU1P3d5DJZO+9996UKVMGDhzYzX8xe3t7Dw8PCwsLAKNGjaqsrNTwJyPaRq+Y9FBRUVFJSYmdnR27kBKH7b5Hqa6u/uOPPywtLSdMmMBVDObm5omJiYsWLUpISAgJCUlOTh4/fjyAoUOHbt++vaqqqqamZljnfQDvu3fv3pEjR9hWZ2lp6X9ewOfzHR0dHRwcBg0aFBwcLBAIRo0aJZPJhELhlStX2FkAgwYN0vgnJFrDdR4n6vdA85Odt3P16lVuo+ps7969AKZPn851IExzc/PcuXMB8Pn88+fPs4UNDQ2vv/76A4untLW1SSSS6OhoDw8PI6Mu7Q8+n+/p6Tl16tSFCxdu3rz5ypUrD531X11d7evrC8DLy+vGjRva+HhEK6g1qoc6Nz+vXr1aUVHh5OTk7e3NdVwddKeBbGJisn///iVLlhw5ciQiIuLEiRMGBgZisXj27NkmJibsNdnZ2f/+97/T0tLy8/ObmprQqb3p7OwsEAhCQkJGjRpl8Li9rRwdHSUSSXh4ONtPqv35VERTuM7jRP06Nz83bdoEYOnSpVwH1YWbmxuAjIwMrgNRaW1tZTe4t7a23rFjB1t448aNd999d8KECQMGDLCxsWHbmwsWLOimvdkTdXV1bFfG8OHDi4qK1PchCGcojeobdgTPoEGD2P/n8+fPB7B161au4+rA9ify+fzW1lauY+mgUCiio6MBGBgYrF692tfX19ra2t/ff+nSpXv27CkvL1djXXV1dWw/7CQfn7biYjXemXCCHur1DTvUKTQ01MDAQKlUpqamsqdcx9VBLBYDEAgEPB6P61g6GBkZ7dmzRywWNzQ05OXlnThxQnNrQdna2orF4iVz5vxffb1hcDAkEmh4PhXRKBo3qm86dztmZmbW1ta6urpqfyWObuhOx+gDTExMnJycFArF3//+d02vp2djY3P82DG+jQ3KyhAYiKtXNVod0ShqjeoVpVKZkpIC4Isvvjhz5oyLiwsAoVDIdVxdsGlUpxrIrMrKytzcXCsrq4kTJ2qjPktLHD2K2bMhlSI0FBIJfHy0US9RN0qj+kCpVF6+fFksFh87dqyurg7AtWvXSkpKBgwYAB1r9+Xm5t64ccPR0ZEdfq9TJBIJgKCgIPWudtodS0skJmLOHIjFCA2FWAxfXy1VTdSH0iiXGhoa4uLijI2N16xZ0z68pudu3LjBbtEuFovZ7AnAxMRk/PjxxsbGBQUFN27cAMAup6Qj2K7bsLCwxw4P0j5uehssLJCQgLlzkZyMsDCIxRgzRqsBkD6jNMqlnTt3zp8/v6qq6ujRo+wr9cdqbW09d+4cu0s7O9eTLXdzc2OXjo+IiDAzM2tra6utrQ0LCysoKHj22WdPnjzZ/YRFrdHZjlF0ejun7YrNzZGYiOhoJCQgJARJSeBuchfpDa6HCvRr7777bn19fV5e3qZNm7q/sn2zTHZ5JJaFhQW7WWZOTk77lVVVVW+88cb27duZTltj6sgmw0ql0t7eHkBBQQHXsTyosLAQgJ2dXVtbGzcRNDczc+cyAMPnM/fnU5E/BWqNcmnMmDGpqalVVVVjHvYc19jYmJ6enpiY+PPPP3eeu+3m5hYVFTVr1qzAwEBTU9MHfsrJySkmJiYjIwOAs7OzVCoNCwvLzMwMCgqSSCTcTuXOysqqra11cXFxd3fnMIyHYpuiAoHA0JCj4SsmJti/H4sXIz4eL7+MS5ege/0e5KEojXJp0aJF+/btGzhwYFBQEFvS/rJILBanpqa2b5Zpb28vEAiEQuHMmTPZjYx6yMnJSSqVsltjhoaGSiQSLW+N2Vl7xyhXAXRDJ3ob2Ey6Zg1Wr1bl0Bs3kJYGhQL+/vDyUl126hT8/GBnpzq9ehUGBhg5kpuYCeihXjdUV1cfOHBgxYoVnVuLPB7P399/7dq1v/76a8+fNJuamlatWhUdHS2TydoLZTIZ2+D19PRU74ScJxIVFQVg165dXAXQDR1cwIWJjWXs7JglS5gXXmAGDWJWrGDYGajDhzMpKR2Xvf46s2YNVzEShiaD6ojODUw3N7eXX345Pj6e3WxDXWpqasaOHQvA29vn+vUWNd65e19++WVqairDMAqFwtraGsD169e1VnsPPTCDVidcvMiYmzPtab26mhk6lGH/AlEa1TH0UK8T5s+fX1xcLBKJRCKRh4eHJqqwt7c/derUjBlRhoa7p00zlkrh5qaJero4ffp0Q0NDaWlpYGBgRkZGfX29l5eXDi5rZHLhwkcTJtT5+urQMKyjRzFnTsejuqMjVqxAQgKee47TsMhDUBrVCf/617+0UAufzz9+PEUk4pWWIiwMUim6mSPa1tb25Zdf1tbWzps3b+rUqb2orqmp6ciRI8uXL2d36eBsOFEPPJWY+P7Fi1i5kutAOikuxgMv4jw88NNPquOYGAwYoDouL8d//ZdWYyNd0Zz6/oXP5yUnIyAAJSUIDsb9LdcegsfjvfHGG1OmTCkuLu5dXRUVFUOGDNm+fXtaWhp05B3OQymVSE0F7m9GrSOsrNDY2KWkoQHW1qrjTz5BfLzqKypK+9GRzqg12u/Y2CA5GbNm4dQpBAZCIsGjpmU2NDQUFxf3upPB3d197dq1MpmspqZGLpezabR9TIIOycxEbS1cXbtrnGvfyJHYtatLydmzHb8qJye4uqqOOw0lJpyg1mh/xK6JERaGqiqEhqLrzpgqCoXixIkT3t7eubm5fanL0dHRx8fn/PnzbW1tBgYGcrm8L3fTiFOnAEDXhmEtWYKyMnz2GVpa0NaGH3/E4cN49VWuwyIPQWm0n2JncoeHo7oaYWHIzn7wAmNj4ylTplhZWb399ttPevOWlpaqqioAzc3Nzz77LICgoCALCwuGYWbPnl1TU6OOT6A+bBrVtd4GOztIJJBKwefDxgZxcUhIULVGHRzQeQUGa2tqkHKM66EChEv37jGzZjEAY2vLXLzYy5u070O8cePG/Px8hmH27t07b948tlAul7MHmZmZbP+At7d3RUVFHyNXG4WCsbZmAIa74bSPx9X8VNIz1Brt10xNcfAgZs7ErVuIjHyCtYMzMpCSAgC1tRgxAkolAOTm5iYmJgKYMWPGmjVr2CsH3H+h7Ofnd+bMGV9f39zcXIFAUFFRoeYP0zsXL6K+Ht7eeJK5YdrG1fxU0jP06+nvTE1x5AjmzYO3N+RyrFyJ9vxWWYnVq7tcHBGhevwvLERcHADY28PeHpcuAcCzzz7r6uoKgM/nP3SMFLs15pgxY/Ly8gQCAbuOH8ekUkDH3tGTPxtKo0Q1kzspCbW12LoVb76pKr99Gzt3drnS0xNJSQAQEYGUFDQ3A8CiRWAXTgkICFiwYEH3dTk4OJw+fXrChAn5+fnTpk3r9WgqtdHNjtFOGIaJi4t744032JG87c18ojsojRIAMDaGpSUA+PsjMxO//PLwy0QiVRq1tYW3Ny5cAID338fChU9Ql62tbXJy8qRJk0pKSgQCQVFRUR+D773mZqSlwcAAISGcxfA47E6lJiYmPB7v7bff5mwBKvJo9CshXRgbIy4Of/sbmpoe8t3QUJw/rxoVfuIEAgN7WQufz09KSpo8eXJpaalAIGDX+uTA+fNoasKYMbC35yYAohcojZIHRUVh1Chs2PCQb1la4uBB1QsPPr9PtdjY2IjF4pCQkLKyMoFAcO3atT7drneCgvDHH/j2Ww6qfhInT57My8u7du3aL7/8kpeXx2X7nTyMAXN/FwpCjh3Dp5/i7FmUlWHsWOzYgRdewM2bGqyxsbFx9uzZUqnU2dlZLBb7aGdrzOZmfPUVjh3DrVvw9MSaNdDBuVWdlJeXt7S0DBw4sL6+XqFQODg4WNFAUV1Ck0HJQ7i4YPVqvPeexiuytLRMTEycM2eOWCwOCwsTi8WjR4/WeK3PPAOZDP/8J5ydkZyM2bNx6BB0bBvqztrXxLKxseE2EvJQ9FBPHu6//xv3l97XLAsLi4SEhPDw8Orq6rCwsN9//12z9WVl4cQJHD6MSZPg6oqXXsLbb+PjjzVbKdFrlEZJBxcXzJmjOjYxwdatWLZMG/Wam5snJibOnj1bJpOFhISwG0lpSlYWxo6Fo2NHiUiErCwN1kj0HaVR0uHuXdjadpyOHQsDA7S1aaNqU1PTgwcPzp0799atW+Hh4RfYsVTqVVCA69dRV4cHHo35fMjlUCjUXyPpHyiNkg7Z2Th0qOP07l3ExaG1VUu1m5iY7N+/f/78+bdv346IiDh37pwabnr3LsRivPMOfHzg4YHNm+HqirKyLteUlWHwYBgbq6E60i9RGiU6xMTE5MCBA0uXLr1z505kZGR6enpv7sIw+O03fP45BALw+QgPxxdf4OpV2NnB0BABASgsRFpax/Xff6/L75eI7qM39aSL1lbcvq06rq/nIAAej7dz505DQ8Pdu3eHh4cnJib2cMH8uzKZRXIykpJw8iSqq9tvh4AAiEQQiTB+PHg8APj6a8yciVdewaBBSEpCXp5q9XtCeoXGjZIOW7fi9dfRvsdzWxtKS3HvHkxNtR1JW1vbiy++uHPnTvY9/qO2tm9ra8vMzBSLxYmJiWXFxWXtq6o4OSEoCFFRmDWrS3dvu+xsJCRALoeXFxYtUs2EJaRXKI2SDlu34uBB1ax5ADIZnJy4SaMAlEplTEzM9u3bLSwsfv75Z2Gn5+7CwsLk5OSkpCSpVNq+nL65uXnuvHku48dDJHrkviiEaAA91BMdZWhouG3bNnNz82+//XbmzJlr164NCgoSi8VisfgSuzAfAMDNzU0oFAqFwunTp7evbUqINlEaJbrLwMBg8+bNcrl89+7dGzZsaH9ysrW1FQqFIpFIJBL1Zdf7oqKi/Pz8yMjIy5cvp6amLl682MnJSU2xk36E0ijp4O7eZeFNMzMsXqx6K8MVAwODHTt2FBQUXLhwwc/Pb9asWZGRkRMmTOCpI6zc3NyTJ09GRkZWV1cvXrx4w4YNmzZt6vttSX9DaZR0EAi6pFFra+zbx1009xkaGs6YMSM8PPzDDz9U751nzJhx+vRpAJGRkTk5OS4uLuq9P+knaNwo+RPYvn37tGnTNHf/K1euxMfHr1q1SnNVED1GaZToutLS0ps3b/r6+qr9znv27Glqajp79uzXX39tZGR08OBBtVdB+gMa8ER03auvvnrixIkHVsgvKipycXHh8XhZWVnOzs7Ozs5chUcItUaJrjt37tzgwYM7l6Snp7/55pu3bt1KTU2trKx87bXXWrU285+Q/0BplOi0M2fOFBQU2HfdK2nKlCnsBs7BwcGNjY22trZGRvSylHCG0ijRaZs2baqvrx82bNijLpg/f76Njc3t9oUACNE6+htOdFdra2t2djaAgICAzuU5OTlZWVm7d+8eM2bMhQsXWlpaaG8iwiF6xUR01/79+5ctW2ZhYXH+/HkPDw+uwyHk4eihnuiuik8/ndraam9jM2LECK5jIeSRqDVKdNWdO4yvr7K6utHc3Jq6PokOo9Yo0VXffWdQXs4zNrbuw+IjhGgBpVGiq44dA8PA0BATJ3IdCiHdoTRKdFJZGfLzMWAA5HKMG8d1NIR0h9Io0UmxsZDJAMDCAlOmcB0NId2hNEp0UloaLC3R0ICBA+HlxXU0hHSHht8T3XPhAq5ehVIJAHZ2oK1BiG6j1ijRPZ99BkND3LsHAF1n0xOig6g1SnRMaytOnEBLi+rUwYHTaAh5PGqNEh2zbRuMjVVP9AC8vTmNhpDHozRKdExaGuztYWYGAEZGmDqV64AIeQyaDEp0j0KBs2fx44/IyUF8PHWPEh1HaZToGKUSLS2q1ighfwb0UE90Rl4ewsNhZYXBg+HlhQMHVOXTpiEhoeOyzz7DypWcBEjIQ1EaJbpBLkd4OAICcOcO6uqweTNWrIBUCgD37qGtreNKhaLjPT4hOoDSKNENP/0EGxusXw92V6XwcLz6KjZv5josQh6Pxo0S3ZCbiwkTYGDQUTJxIuLjVceHD+OPP1THv/4KWjqP6BJKo0Q3NDfD2LhLiZmZaiITAGNjmJqqjnk8rQZGyONQGiW6wc0NBw92KcnNhbu76nj2bMybpzpubERxsVZjI6Rb1DdKdENUFM6fR2qq6rSuDt98g+hoTmMipEeoNUp0g4sLtm3DvHkICoKVFaRSTJ+OmBiuwyLk8Wj4PdEldXU4exbNzfDzQ/uOypmZGDYMAweqTsvK0NwM2m+Z6AxKo4QQ0ifUN0oIIX1CaZRwrLGxcdOmTZcuXWJPv/nmm/ZjQv4UKI0SjikUCg8Pj2vXrgFITU0tKysrLCzkOihCngClUcIxPp9vb28PoLm5OT4+fuHChVxHRMiToTRKdEVhYaGtre2uXbvS0tK4joWQJ0DjRgnHGhoakpOT29raIiIiPvjgg/Ly8rq6Oq6DIuQJ0IAnQgjpE3qoJ4SQPqE0SgghfUJplBBC+oTSKCGE9AmlUUII6RNKo4QQ0if/D2Yz2BYHnSFNAAABQXpUWHRyZGtpdFBLTCByZGtpdCAyMDIzLjA5LjUAAHice79v7T0GIOBnQAB+KG5gZGNIANKMzGwMGkCamYUDQjOxM2SAaUY2BRMgg4mRBcJgZAQqAWtByKAKAFVADGVigklwMCiA+BgmcDMwMjAyaTAxMiuwMLMwsbAysLKxMLKyM7BzsDCxczJwcrEwcnEzcPIw8PAy8PKxMPKyMIgwArWysfDycLKziu8DGQP31GfbqQfiBc+ABBmE//seWKCQbA1iK3GaHfgZu8sexDaq4DzwgaVtP4jd8NZ4v6CcKphdqb1i38FdfmA1CyQ27394+S2Yfexdn73n5Vlg9opb++xvxms5gNgu/7gd7J8p2YHYYtUZDglL+cFq/jRMc3Du/wV2wzKBF/YL+5eCzeeI2WPXyXoTzGZ5cGX3zwrPA2C9AIxhR2+Z1+LoAAABp3pUWHRNT0wgcmRraXQgMjAyMy4wOS41AAB4nH1TwY7cMAi95yv4gbHAgA2HHnZmdldVtYnUTvsPvff/VfBoNpnKqu1YDrxgeLwskOP79dvvP/A56nVZAPA/y93hFyPi8gF5gPPr+9cVLreX88Ny2X6utx9AmgtzPmNfbtvHw0JwgZOU1tiowgkLu2mPTwqOsX9bE8mlkjWX4WdpbQJk2OBUi4kwdqAiqu48AQqsCSTp3ghOVFA61llIzbuxhNeE89RE3GdZtjuSA6g9/IZduU+APbOkKNfJPLI0Naw+AVpEpELUWqtxqOjaZ9V4BIxinVAMajBaCWcZUjQo/GSdB4+kjkYzYPaGS+OqlBdqI6JZKdG4DSTY4Sqa1Ys5d5kheSRpFAKyLL86i8wYJ4nLg8fWLXgOZHOVeZqaTKa9ibeUCIkhzqh8Xa9P4rvL8byt112OOeuuuZy8Kyt8ILt+KJQRib69f+FdKmlsux4onj4wtPc+jbZ3mOLx5zg+jHRsG42NnmPRSDlWPbTjjuQD7TQ2OdBLY9N/oundrkfKjgTl++Nfj/PyFwVuzOczflG1AAABR3pUWHRTTUlMRVMgcmRraXQgMjAyMy4wOS41AAB4nE2QsWoEMQxEfyXlLXiNxpJly0vgYFOkyqZJFVKkSBkuhINr7uMj3xG8lTWDNfPsdT08btPL+3o8Pn/gfhy2yYf/sw/rNm03C9vD9TBLVOWKMFNkq7mEZeaYUNUCRSIWzT4uc4pVhBEQJWe7O5BiGmZEkkLsFsVCqQr3NBW5XfNctzyYYqWSuTfAuwzmYTXXvokIqKobiaxDeI2BckiOlxAWL6uFOxOykfMuHJVThjtZAfiOeDsn6eVSjYv0mAomdMZkLDksjqOlWnfUstQbIZGKaedmf1T1hSl8nk/fr7+nn0axjy+n81ekhiHQ0hCp8RDcZAhpeYjcdAhtZYjS6hC12RDWsEOAM+wh0LDDQGrYgYAbdiiQcHlznBws+uf3t16enJVDieX6B3k8kURS7OVYAAAAAElFTkSuQmCC",
      "text/plain": [
       "<rdkit.Chem.rdchem.Mol at 0x1ca05f5a0>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.input_ligand_templates[\"1.C\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcIAAACWCAIAAADCEh9HAAAABmJLR0QA/wD/AP+gvaeTAAAgAElEQVR4nO3dd1RU1/o38O/AUIYOgyigKKAYRbkSTNQo9qgvisRErJEQYwEvYL96bRj9qZjEBLghRgRMjA2xF4SAsXOjSFFERBCwASIBBIahzMx5/xguCKKhDBxm5vkslovZZ59zHlj6uM/ZjcMwDAghhLSWCtsBEEKIfKM0SgghbUJplBBC2oTSKCGEtAmlUUIIaRNKo4QQ0iZctgMgLHv16tWhQ4dEItHcuXONjIzYDocQ+UOtUWW3cuVKPp9vZmbm6urKdiyEyCUODb8nUgMHDkxJSWE7CkLkD7VGCQBER0d/8MEHbEdBiFyid6MEt27d2rlz54kTJ9gOhBC5RK1RZRcSEuLh4bFp06ZHjx6JRCK2wyGy8PQpEhPx5EnjcrEYRUWoqmpcLhSiqAj0fq+1KI0qOqEQ16/jzBlcuYLy8jeP5+bmTpgwISoqKiIiorq6uuMDJDLDMPjxR1hYwMICDg7o2RPm5ti1CxJJbYV798DnIzi48Yk7doDPR15eB8erMOihXnGJRNi4EYGBqKioLVFXx4IF+O478Hh1tTZt2sROeETmVqyAvz8+/RShobCwwLNnCA7GqlVIS0NICNvBKTJKo4rryy9x4AC8vODlhe7dkZ+PkBDs3InMTFy4ABV6EFEscXHw98eXXyIsrLakb1+MGwd9fezdi5kz8fHHrManyOjfkoL64w8cOIDly/Gf/6BvX2hrw9oaO3Zg82b8/jvCw9mOj8javn3gcLB1a+PyrVvB5dbnVtIOKI0qqIMHweHgX/9qXL5iBXg8HDjARkykPd26ha5dYW7euLxrV1hYID6+vkQkQlVVgy+xuCMjVTz0UK+gEhNhbIxu3RqX6+jAygpJSWzERNpTYSG6d2/6kIUFEhPrP65YgRUrOiYoJUFpVEEVFzfRMJHq3h2ZmR0bDWl/amoQCps+VFEBDY36jwsWwMmpQYUjR3D0aDvGpugojbbJ3bt3q6qqHBwcVDpbj42mJkpLmz706hW0tDo2GtL+evTA3btNH3ryBD161H+0s8O0aQ0q0NNJ23Syf/xyZdu2bQEBAcePH1+4cCHbsbyhVy88f970O6/Hj2Fp2eEBkXZTUwMAo0ejtBQ3bzY+mpyM/HyMGdPxcSkPSqOtxDDMoUOH9u7d6+fnl5GRUVBQwHJAL18iK6v+48cfo6oK0dGNq928ibw8GvuiIDIyMGMGvLwAYOFCaGvDx6fBJIuKCnh5QVMTnp5sxagMKI22UnFxMZ/Plz7LW1tbP378mM1oHj3CiBGYMAF12fzLL2FiAh8fPHtWX62wEIsXQ0+v9h8eYVtCQsLKlSulk8dKS0s3bNjg7e394MGDvz+ztBSrV2PAAERE4PhxlJTAwgKhoUhKwoAB2LwZe/di61YMHIibN7FnD6yt2/2HUWL0brSVdHR0BAKB9PuysjI9PT3WQrlxAy4u+OsvDB5cPy3ayAgREZg6Ff37Y/p09OiBvDycOAGhEEePvrVLl3SgtLS0mJiYixcv1tTUqKurL168eMaMGba2trNmzbp16xaX+5Z/mwyD337DmjXIz4eKCubNw7ffwsAAAGbOhI0NvvsO+/ahuBgGBvjoI4SHY/Dg2nO1tDB4MLp2bXxNc3MMHgx19Xb7WRUdQ1pr7NixKSkpubm577//vkgkYiWGvNOnGU1NBmCcnZny8saHc3OZdesYR0emXz/mo4+YVauYnBw2wmxf8fHxrq6ukydPvnjxItuxtNjYsWPLy8sZhunfv7+0ZN68efHx8U3XvnWLGTqUARiAGTKE+fPPDouTvAO1Rltv3759mzZtqqmpCQ0NVVVV7fgAAgIC1q5enWNv39XODrt34832i6kptm3r+MA6UkVFhaen54ULF4yNjRm5XaNIJBKp/68xaGRkVFJS0rhGbi7WrsWBA2AYmJtj+3bMmwcOp6MDJU2hNNp6FhYWv/zyCyu3FovFS5cuDQoKUlFROT579pJlyxocPnYMhYXw8GAltg52+/ZtGxubffv2aWpqLlq0SOP1AZLyg8vlikQisVisqqr6+PHjHq+NTxIKhYGBgWYxMfMuXgSPBx8fbNgAHZ0WXV8gEGhra9ddUENDo9MN0ZNn9KuUPwKBYNq0aUFBQRoaGgcOHGicQ3ftwsyZ8PJ66yhCxVJcXPzs2bPx48erqKisXbuW7XBaQCQSRUREFBQUnDx5sqioaPr06atXrw4ODq6qqurbt6+0TkRERL9+/dauXfvlpUtFnp548AB+fi3KoZmZmXPnzjU1NZVIJDU1NStWrLC1tY2MjGyfn0lJUWu0TcrKyi5fvszj8bS1tdXV1fX09FRVVQ0MDFRVVfX19dvjjvn5+c7Ozrdv3zYyMjp16pSjo2P9MbEYy5bhxx/B4WDjRtjZtUcALfXq1SsAdb8NiUSSk5NjamrKe22xvrawsLCwsLCwt7fX09NbIYdzHOsWKvT19Y2NjX316tWxY8cApKWlLV++PDo6GsCgQYMCAgKMRo5sxfVNTU3DwsJGjx4NQE1NbcOGDQbS/igiO5RG2yQnJ2fq1KnvqKCpqcnj8Rr9aWho+GZhc6oVFBRMmTIlMzPTysoqMjKyrs0CQCAQHF65ckFYGDQ1sX8/Osc2n1u3bj19+rSLi8vGjRulJXv27PHz8wsNDR0/frxMbmFvb19eXr5s2bKEhIS6u8gFLpfbaDdW6e+kqKjo3//+d1BQkFgsNjIy2rRpk5eXV6tfvtc9y0vRHtrtgdJom2hpaU2ZMqWysrK8vLympubVq1disbikpEQsFpeWlgKorKysrKyU7U25XG5ubu7AgQMHDhx4+PBhGxubFy9eODs7x8fHa0ycOG/jRgwfLts7ttq//vWvQYMGJScnSz/m5uaeOnVq9uzZsr3L8ePHHz161K1bN11dXdleuYOJRKKwsLD169cXFhZyudxFixZt27bN2NiY7bjI36A02ibW1tZnz559RwWhUFhZWdnoz+Li4jcLm1OhtLRULBaLRCLppklpaWnDhg3btWvX9u3bMzIyrKysPgwIwGtNVNY16vDx8fHZuXNnRESEbO+ioqLSp08f2V6TFefPn1+8eDGAiRMn+vv7v/fee2xHRJqF0mj74vF4PB7P0NBQVhcsKysTiUT3798vKSn5+eefz507N3/+fIZhhg4deubMmS5dusjqRjJ39OjR8vLywsLC7OxsQ0NDWT3UK5KpU6d+8cUX06dPnzJliqyuWVZW9tVXXwkEglmzZu3ateunn366fv36jRs3JBLJu99HkeajNCpnpM+tw4cPB+Dk5PTNN9+sW7eOYZgePXo0egvW2ZiZmX366adZWVklJSX5+flsh9NZJCUlXbhwwdvbW1dXl8PhyHwIna6u7tHXFsHbsWOHbK9PAKhu3ryZ7RhIK3E4nBEjRgwePPj8+fOJiYnnz5+fNGlSp+qHDQ4OPnToUFZWVnFx8YwZMxwcHBwcHEpKSiZMmGD+tuVQm0Msxs2buHoVKSmoqICpaYOtpe7eRWYmLCwanCKR4NIliETg81t/33ZQWVl59OjRZcuWSSQSe3t7dZqRKY9YnkVFWi4mJmbixIm+vr51Jenp6f369TMwsBw0SCiH8yFbKDKS6dmTARgul+FyGYDp1Yu5cKG+gpMT061b47OEQgZgvLw6MtLmS0tLmzdvXteuXX19fV+9esV2OKRlaPi9/DE0NFy9erV0PKaUjY3NjRs3pk+/lpysOWkSgoJYjK6dXbqEqVNhbIzr12v3Ebp2Dfr6cHbGpUtsB9diUVFR8fHxAN577739+/dfvHjxwYMH7C+6SFqI0qj8cXBw0Hpj+XpDQ8PgYHNfX4hE8PLCvHlv3VFCjjEMvLzQpQtiYzF8OFRUoKKCESMQGwtDQ3h7Q97m1JeUlHz22WfOzs6JiYkAbG1tjxw50rt3b7bjIi1DaVRxcDjYvBmHD0NbGwcOYMQIPHnCdkyylZKC+/fh7o5G73+NjeHujtRUpKSwFFkrzZo1KyMjY9KkSS4uLp988klqairbEZHWoJ56RTNzJvr2xbRpSEzEqFF48ADtu1jH8ePQ0sL/+38NCouLcfIkhg5F//71hSIRLl/GnTsQCGBignHj0NLBnnfuAEC/fk0cGjCgtoJ0CmxNTePtNKqrW3av9hcbGzt8+HAej/fPf/7zq6++Cg4OzsvLs7W1ZTsu0nJsv5wlLXbo0KGxY8f26dNn69atb6tTWMiMG8cEB7d/NH36MCNHNi68c4cBmICA+pKkJMbGhgGYrl0ZGxtGS4vhcBh3d0YobMG9AgMZgImNbeJQbCwDMIGBDMMwTk61K3K++dVpupgkEsmMGTPMzc0DAwOFLfolkM6HHurlz+zZsy9evPjw4cMNGzbUFSYlITy8fgs7Ph+//46FC3H2LK5erS08dQrHjkEkanC1e/cQEdG4UMaeP8f48aisxJUryM9HejoKC7FpE379FV991YLrSFczKS5u4tBffwFA3cjZLl2Qnd3gqzk7c3QgDocTHh5+9uzZ2NjYPn36/Prrr2xHRFqP0qiC+OUXzJqF16fvS0dSenigbsD1F1/A1RWBgQ1OPHwYM2a0c3/U9u0oKsKJE6hbo4jHw+bN+OorHDqEhITmXkfa99Lktlc5OfUVAKiqolevBl89ezbzJiKR6PDhw2PHjs3IyAAQGxvr5OQ0ZcqUM2fONDfOv8MwjHTDJXt7+9OnT586dUreVwNQcvRuVLmoqcHXF66uDfYtb3cnT+L999GzJ3bubFAufct58iQcHJp1nSFDYGCA48exYkWDhd8ZBhER4PPx4YdtD7aiokJbW7tHjx5lZWUAnjx5cvToUQ6HM3jwYFnNnszOzh4zZszIkSN9fX379+8vnZUgkysTVlAaVS6ffYbYWPj44OTJJo6uW4f9+2FkhN69axtwvXqhe3eYmKBLF2hqvuWiL18iPLxByetDBIqKkJeHoUOxfTt++KFBNWkqbH73NI+HlSuxcSO2bsWGDbXtbYkEmzfj9m3s2PH2EFtAT09v6tSpFy5ckH6cP38+gJqaGhmuq29lZZWZmfnTTz+NHTt2/Pjx+/fvp7Xo5RqlUeViYIDNm+HlhTNn8GbT6ulTPH+OvDxkZ4PDgUiE6mqoqUFPD5qa0NSElhZ0dcHjgceDjg5MTLChSGT0VwZmzXrrLcvKAEBbGz/8AGPj+pehMTFITISGRm2FZlq7Fg8fwtcXR47UviK4cgUPHsDdHatXt+A6LbR+/XoPGW3KUlFRweVytbW1V69e7enp+ccff1AOlXeURhVKdHTj4U1VVY3rLF6M4GD4+GDcODRazETaOpRIUF5eW6KmBg0NVFejshICQX0XVp1/quAe8LG6Oo/H09DQ0NbW1tXVtQN+u3s3ODj40n//21NHxw/IS001BWqMjES+vpqamgcPHvy8vByJiaiqQoteC3K52L8fbm6IiEB6OgCMGoWgIIwdW19n0qQGA63qTlywAB991IJ7AQDEYvH69eu1tLRklUbDw8O3bNmyYcMGNzc3HR0dWmZJEbA9VIDIho/PWwf5TJpUW0dPj/HwYBiGuXqV4XCY1asZhmHWrWMAprSUYRjGze2tFwEYDQ1GV5fR02N0dBgOp7YwU8Xyyht/qewABvABVFRU9PX1X6qopKqqMsBDdfUxY8ZIV6JKGTWq9hLr1rH0O3ur7du3Dxo0yN3dPTExcfLkyQMHDlyzZs2aNWvKyspkcv24uLiPP/7YysrqyZMnMrkgYRe1RhXKixdoNE3UxqaJao6OcHPDDz/giy8alL97v17pFHYpLS2oqIBhwFSoqKmpWZiaamlp6erqamtr6+npvVddjaio6dOnf7JkSbdu3fR37DA+eBBAnz59YmJipPtYXL16dYD0Ws1vjp0/j1OnMHEixo2D7JZwfdPcuXNnzpwJwMTERLqZh7S87ftHSff+HDZs2O+//37r1q3u3bu3NVbSCVAaVSja2k0/p7/pm29w5gxWr4a9fX1hk+/o1NWhpwctLWhqQkcHenrQ1oa+Prp3h40Num1E9x726VeuaL7evXP3LqKiHB0dMWYMAGzahGPHIBTi4UNVG5sSPv+JQGAiFgOosLbWGjKkuT/eP/6BrKza0ab9+2PiRLi6oh2m/Vi8tsieTgu3Mn63OXPm8Hi8jRs3WltbfyiLcQWkM6A0qqRMTLBtG5YsqR23LmVjA0dHaGnBwAA6OujdGzY2sLREt24wMUHTm6rtVKntfnqH3r0REYFp01BTg6wsDiAdw1kJjH76dF9qanNnQHbvDm9veHujuho3buD333HvXnuk0fYTHBwcEBAwbNiwvXv3uri4sB0OkQ0OI2+L4pAmLV2KwECUlzdujZqbw84O0tE7+vqYMwe7d9cekkjw0Ue1U89LS1vW01Pv55+ho4PPP29QWFCAvXsxcSIGD64vfPECoaFITZXOqX/Sq9f7vr5/iUQ6Ojrx8fHN2ndIImm6wSxvSkpKuFyubNu5hE1sv5wlsiHtYiovb1xuZtZEF1Od27cZVdX6LqYOdubMGQ6HA8DIyCgrK4thGObyZcbWlhGLmz5h+HBm1Chm+3bm9u231pGVqipGJJLh9SorK8eMGXPw4EFxe0dOOhxtIqIgJBL07Inx4xs/eotEGDoUgwYBQGUlhg9vsECSmRnMzdGvH8aPB7fDX/D07dtXQ0Pj4sWLQqEwPDzcVVvbYOVK5OXB0BAffNDECbNmoUcPpKTgu+/w/ffw9v6bTrFWyMvD2rVYuBBr1mDLFhw6hJcvMXhw25fJ4nK53bt3//bbb1NSUiZNmiSTYEknQQ/1hGULFiwIDQ0F0EdV9YpYbKqnh169EB+Pd+9KVFws+876R48wciTKyuDpiaFDUV2N2Fj88gsGDMCVK9DTk8lNBAJBJ998kLQY281houyqq6oOmpsPAgAMBAq1tRkVFWbz5sb1KiqYu3fbN5RRoxht7cZ3OXaMARhPz7ZcOD09fcmSJU+fPm1TeKSzUoQX9kSuqd2/P1ssjgFsgRRgvFBYLJHgyJHGk0Szs/HppzA3x/z5uHtX9nGkpuLKFXh4YODABuWffYaPP8Yvv0AgaPW1+Xy+rq7unDlz2hok6ZQojSqL9PT0rKysjr9vfHw8n8//7bff3lpj0CBObKyunV0M8B6QLJE4AWXp6Vi/vkG1/v2RkYGrV/HBB+3SX3/9OoCmZ4s6OkIoRFJSq6/N5/P9/PyuXHlzwhdRBJRGlYK7u3tISMi///3vDu5RFIlE69at8/b2/pt6trYaCQnM9OlHOBxL4E9gEofz06+//vX6oFYpa2t4etZuGSITEgmePgXDIC8PQNPLkkoLpRVaLjo6+tSpUwzDcGTeIUY6Bxp+rxS+++47Y2Pj/Px8Nze3jrzvzp07582bV1pa+vdVuVyziIj7K1eGff/9F0CcRBJXWvrn8uX79++XHo+KihIKhePGjdNrdVdPcTGysuq/cnORl4e0NFRU4MWL+nVZ3iTthm1tEtTQ0PDz8xs5cqR0FixRPJRGlYKxsfGuXbt+++23gICADrvpgwcPYmNjjx8/Hhoaqqam1pzm2Phdu7YAtt9/L12v9NixY76+vtbW1gAEAkFISMhPP/0UExPz7osIS0p42dnIyandPqTum4qKpk8wNcXLlzA1BYCnT5sYayVdPtXM7O9+4qaNHj169OjRrTuXyAUa8KREkpKSli9ffvny5Y65XWRk5OnTpwGkpqaqq6ufOXOmOfN2GIaZM2fOkSNHpB/Nzc1zcnK4TQ1qraqqev78eVZDubm5epqaD7Kzm7i0oSFMTWFmBiur+q++fSGNKjUVAwZg1Sp8+23jEydOxPXrKCzEX39h5kx4e+Ozz6Cm1pxfwrlz54YMGdKlS5fmVCZyilqjio9hmMjIyMmTJ1taWpbXrSTa/pycnJycnACEhITo6ek1c+4jh8MJCwvLzs6+efMmh8N5/vz5sGHDQkJCCgoKsrOzc3Jy6v588eJFk1eoNjJiBg7k9OoFS0tYWqLum3e/DbC1haMjfv4ZCxagb9/68gsXEBODJUvA42HPHsTFIS4OPXrAywuLFsHA4N0/zvXr183NzSmNKjZqjSqF0NDQ6OhoiUSybNmyESNGsB3O38vLyxsyZMjTp085nLf+FVVXV+/evbuVlZWVlZWpqamZmZn0e0tLy1Z25qSnY+RIiMVYuRLDhqG6Gr//jqAgWFvjxg3o66OqCuHh+PZb3LsHADo6mDMHy5Y1mBlGlA+lUdJJJScnOzo6lpeX6+jocLlcCwsLOzs7GxubXr16WVpaWlpampmZyb7vOycHGzbg9OnaDQBMTDBnDnx9G7Q6GQbR0fjhB8TEgGEkxsZzRo709PEZNWpUo4vduXPnH//4h4wjJJ0PpVHSeUVGRk6dOlUikezdu/erFu1o30ZiMV6+hKoq3v0wfu8e/P0v5eSMvXgRwKBBgzw9Pd3c3KRLrxYVFcXFxU2ZMqVjQiYsojRKOrVdu3atWrWKw+GsXbt2+/btbIfThJcvX4aFhf3nP/95/vw5gK5du3p4eHh7e/P5/Lo6Eolk/fr1aWlpfD4/MDCQ5tQrGEqjpLPr379/WloagJ49e168eFE6/qmzqaqqOnjwoL+/f0pKCgAej5eUlNT3f11VUVFRJ0+e3LNnT1BQUE1NzbJly1gNlsgYzWIind3du3cnT57M4XAeP35sa2u7Zs2aZo3n71gaGhrz58+/e/euv78/h8MRCoVxcXF1RysqKqRj73v37p2QkMBemKRdUGuUsObVq1e6urrSXdoLCgoYhunatevbKt+5cycoKCgsLEwsFvP5/I0bN3p5eak2vbEJayIiItzc3CorK3v27JmUlGT4v6X8qqurly9fXlhYqKampq6uHhYWxm6cRMZYWVeKKLmCggJ3d/euXbtmZGQwDLNu3To3N7dp06atWbPm3ScmJCSMHDlS+le3X79+UVFRHRJvs/j7+0v/S/Dx8XnbEvcBAQFBQUEdHBhpb5RGCQuqq6srKirmzJkjTaPZ2dnSQgcHhybrX7p0ycXFxdXV1c/Pj2GYM2fOWFpaSpPplClTMjMzOzD2JohEIk9PTwCqqqqBgYFvVhAIBO7u7h4eHs7OzgKBoOMjJO2K0ihhTV0aZRjmwIEDo0aNOn78eJM1/f39Gx2qqqry9/eXLlOipqbm4+NTUlLS7hE3pby8XDqqSUNDIzw8/G3VCgsLpS8uiOKhLibSKXzyySfffPPNt2/OZwcA5Ofn37hxw8/P74l0lRBAXV196dKlDx48WLRokUQiCQwMtLa2DggIEIvFHRg18vPzR48efe7cOSMjo9jY2BkzZrytJp/PpymhCovtPE6UV11r9Nq1axKJpLq6esCAARKJ5M2amZmZycnJcXFxDg4Ob1Z4/YWpvb395cuXOyJ6hsnIyOjTpw8AKyurBw8edMxNSSdEPfWEHS4uLjk5OYaGhjt27Lh37150dLRIJJo7d66rq+ublZn/LbI3bNiwyMhIw6Y2szt79uzSpUuzs7MBTJkyxd/fv11HmP7555/Ozs6FhYUffvjh2bNnTUxM2u9epJOjNErkwKpVq0QikVgsFolEu3fvflu16urq3bt3b9q0qbS0VE1NzdPTc8uWLfr6+jKP58SJE59//rlQKHRxcTl06JCWlpbMb0HkCcutYUKaQSKRPHz48NGjR82pnJubu2jRIumQUj6f7+/vLxKJZBhM3cCmhQsX1tTUyPDKRE5RGiWKKSEhwdHRUdpWkNULU5FIJN1XisPh+Pr6tv2CRDHQQz1RZI1emAYEBFhZWbXuUlVVVW5ubkePHtXQ0Ni3b9/s2bNlGimRY5RGiYJ7/YWpurq6h4fH1q1bW7ovXlFRkYuLy/Xr1w0NDU+dOlU3MIAQUBolSiI3N/frr78ODQ0Vi8XGxsYbNmxo/pT8rKwsJyen9PT0Xr16RUZG9qO17klDlEaJEklISFi+fPm1a9cA2Nvb//DDD28uWd/IrVu3nJ2dCwoK7OzsIiMjzc3NOyRSIk9oFhNRIg4ODlevXpVOyU9KSho9erSzs3NWVtbb6p8+fXrMmDEFBQUTJky4du0a5VDSJEqjROk4Ozunpqb6+fnp6uqeO3euX79+S5cufXMN05CQkOnTp1dUVMyfP//cuXMtfZ1KlAi7AwUIYdHz588XLVokHQRqbGxcN8JUIpH4+vqCBjaR5qF3o0TZvf7C1MLCwtvb+/Tp09evX1dXVw8NDf3888/ZDpB0dpRGiZIqKSkpKiqSfm9oaBgdHb1y5crc3FxpCY/HO3fu3NixY9kLkMgNSqNESV2+fDkqKgrArVu3ZsyY4eHhUVRU5OrqevnyZS0trf3790+bNo3tGIl8oDRKlN2ECROOHj1qYGAg/Whtbf3w4cPOtssT6cyop54otWvXrtna2tbl0GvXrlVWVlIOJS1CaZQoNX9/f+lqI3UfBwwYwGI8RB5RGiXKKyMjQ11dvW6xEqFQmJKSYmpqym5URO5QGiXK68KFC8uXL6/7uG/fvqysLHt7exZDIvKIupgIqTV69OibN29evnx5yJAhbMdC5Am1RgkBgGfPnqWnp/P5/L59+7IdC5EzlEYJAYAdO3bk5+cbGhrW9doT0kyURgkBwzBxcXEA+Hw+27EQ+UNplBDExMRkZGSA0ihpFUqjhODHH38UCAQAevbsyXYsRP5QGiXKTiAQ3L9/HwCHwxk+fDjb4RD5Q2mUKLuL//d/A5480QQMDAzs7OzYDofIHxo3SpSd5KOPOPHxhTxeMcNYFRdzuVy2IyJyhv7GEOWWlaWSmQmRqEtFRRcdHVAOJS1HD/VEuW3fjpcvAUBTEx98wHY0RC5RGiVKTCLBrVsAoK2N6mqYmbEdEJFLlEaJEjt/HpmZ4HCgooKaGgwcyHZARC5RGiVKbPduCIXQ1kZZGdTVMWwY2wERuURplCirV6+QlgYeD6ABRZAAAAN6SURBVBIJABgb47332I6JyCVKo0RZ/fgjHj8Gl4uKCgAwMADNBCWtQmmUKKvz56GtjfLy2o/GxqxGQ+QYpVGilNLScO8eANRNP6GmKGktGmxMlNLWrZBIIBDUl1hYsBcNkW+URonyEYlw8iRqahoUDh3KUjRE7tFDPVE+wcHgciEW15cYGIB2siOtRWmUKJ/ERHTrBh2d+hJDQ/xvm2VCWopWeCJKSSxGQgIOHkRKCp48AY+HlBS2YyLyitIoUXpVVcjLQ69ebMdB5BWlUaJ8qqpQUQFDQ7bjIAqC3o0SpSES4fvv0a8fNDVhZAQtLTg74/bt+gp790JHp4mneycn0Kr45O0ojRLlIJFg+nSsXAk7O4SHIyoK27fj3j2MGIHz52vrVFdDIGjQgy8lFDYYYUpIQzRulCiHkBCcPo1t27BuXW3JxIlwc8OQIXB3R1YWdHVZjY/IMWqNEuXw888wN8eaNQ0KjYywZQsKCxERwVJYRBFQGiVKQCDAnTv44AOoqjY+JN1ROS6u44MiCoMe6okSKCiARIIePZo4ZG4OLhf5+fUl48Y13tiuuLjpcwkBQGmUKAUVFQBN9B0BkEjAMA3y5oIFMDdvUOfHH5s+lxAAlEaJUjAxAZeLp0+bOPTsGcRimJrWl8yejUGDGtQ5eRJPnrRvhESe0btRogR4PLz/Pm7daryqE4CrVwHA0bHjgyIKg9IoUQ5LluDFC2zb1qCwsBBbtsDMDJ9+ylJYRBHQQz1RDm5uOH8eX3+N1FS4usLAAPfvw98fBQU4exaammzHR+QYpVGiHDgcHD4MR0fs2YOZMwFASwsTJuDrr+snemppwcQEamqNzzUyqt32jpCm0NIkRPnQ0iREpiiNEkJIm1AXEyGEtAmlUUIIaRPqYiJK4b///e+JEyfMzc0XL17M4/HYDocoFGqNEsVXWlq6ZMmSxYsXV1dX79y5k+1wiKKh1ihRfBoaGkKhUCAQqKurG1IHPZE16qknik8sFru7uwuFwuTk5FOnTg0YMIDtiIhCoYd6oviuXr2qr69/7NixCxcueHh4sB0OUTT0UE8Un6mpaXJycnJy8p9//mlHm9MRWaOHeqIUkpOT//jjj27dus2YMYPLpdYDkSVKo4QQ0ib0bpQQQtqE0ighhLQJpVFCCGkTSqOEENImlEYJIaRNKI0SQkib/H89aJCUNQIbVgAAAW96VFh0cmRraXRQS0wgcmRraXQgMjAyMy4wOS41AAB4nHu/b+09BiDgZ0AAfihuYGRTMADSzCxsChpgmgNCM7ErWIBoRmY2BRMgg4mRBcJgZAQqAWlhZILLoAowgpSCBJjgEhwKCiA+hgncDIxAUoOJkVmBhZmFkYWVgZWNhYmVnYGdg4WRnZOBk4uFiYubgZOHgYeXgZePhZGXhUGEEaiVjYWXh5OdVXwfyBgoZuB/od16gL3Ay/7KduP92VXaB4RW2dgzbi7e/34+14FjC6bbv1j4Z3/tjp/7uTZJ2BVPi7H9fOvrPnNDZbv0Bk/bJTzMdmw+/Pvb50zbv5LzpF1Y6Rm7jWW8B17c22a/7Nay/e8vhe1/u6LL/uNS9QNTDPfvu7iI3yExRGx/95EGu9xySwedKyb2Pg5N+y4fKHTYtmq2/YoQB3v3n/vtSw8279sru9M+5fAB27Pf7+/r7VtjL7329z5JlTQ7yR+yDmIAWD1mMInvfXUAAAHielRYdE1PTCByZGtpdCAyMDIzLjA5LjUAAHicfVRLbhsxDN3PKXQBC/yK4qKL2E6DosgYaN3eofveHyXlOJLToiONoKEeH59IYraSz7fz11+/y/vD520rBf4z3b38ZADYXktuyvH55cteTten491yuvzYr98Lak7I8Yh9ul5e7xYsp3KQig0beIFqDmZUDrEDApLpS4mk2sxZLJHc3CCRrui6ROFySSS2BkgFK3ZRl3LA6umEEyllH2YFFw5OFEdMclDuugTXDA5VOmBvCVTnNoDh4XnL0+eXT3SHt4AHiMytp5+qmw4FaK0pveHxjrdQDJWdxllsnJnzDoQQiqeOHsRYhXpLYqwMrDJCdHbU9lGIBzFW6EwdRvYyU5x4NhbSyYzhWSKgCOMAaDdKQCVFBfgoGXHgu5ujZDmgcRuOpMJuCzOFCK5mxlEFjDs1vFVQkZwWIA+14sGpNyZj1jRJI1oqHBEzwVE1pBFT2GIGkqXjP8RqtkTAIkc67kRioYGqtE6tT+bn/fzQn7eOPV728+zYHDTbMgfP3ouzIrPBMHpHHtXoMLbZLBivDQzPjkhjn3XHeP2Rx4cR1xLiWPCRC4fkmLSU5IbkJfc4FllyjGPRv3I57LqmbE1Qft9/B7Hf/gCFQdY+IxIR3wAAAXx6VFh0U01JTEVTIHJka2l0IDIwMjMuMDkuNQAAeJxNUD1LBDEQ/SuWd5ALme9MFkE4CyvXxkosLCzlRA5s/PHOrEiWXch782YyL+98Ptyux8eX893dwyv8HYf1GOD/THBej+tWgvXm53DiChpfadW8mWE5BWrYkJcTVjUnSY3ULSUXcEkFVBsUqNBZvJygevYsCaQ5xwywA1iAJtSF475WuTfomqI46SZGl7MtUUJz67lExE04LwVTldDIUQVLAieiEgYQGtkClbFrz15qJJLjnRxEQ2qdsGevhulUyIhRlhhmJthWdUPMF6JACt3NtzCaklJ2oDC5LVTNjDxejBHXlokAOqYDj6G/VqOIKyqsiGE7ggDELTgmiz804g5bFGFSts3Ixlawsna0Y3m7Xj6evi6fo9WEj5fre20DJoGBk+CgSWjwJDxkEhk6iQ6bxEafpA+fxAfsLEB42JuAATsbgAN2RoAG7KwAl+/n8EbFamRAUfi+D3tSvPrPL9+ZqWFTzS/EAAAAAElFTkSuQmCC",
      "text/plain": [
       "<rdkit.Chem.rdchem.Mol at 0x1c9de46d0>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.input_ligand_conformers[\"1.C\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcIAAACWCAIAAADCEh9HAAAABmJLR0QA/wD/AP+gvaeTAAAUp0lEQVR4nO3deVCTd/4H8HcIIWgUFTQMHgQvEBStrMcWqdR69BhRa1ftrGerhVqrdqosaKdqp+7WolvFsc7oWruK7LgwZZSVVSur06JYRS2Xyx1QTq0QrnCEHL8/korbXw8geXgivF/DH0/Ck+/388wwb77fPMdXYjKZQEREXeUgdgFERE83xigRkVUYo0REVmGMEhFZhTFKRGQVxigRkVUYo9RF165dy8rKqqqqOnbs2P3798Uuh0g0jFHqIjc3t7S0NJ1O9+abb544cULscohEwxilLho2bBgAT0/PhISEkJAQscshEg1jlLooNTW1uLg4KSkpOzu7tbVV7HKIRCPhzaBERNbgaJSIyCqMUeqwjAy8/nr7S19fy8b+/ZgxA7NmYc4cZGaKUhqRiBzFLoCeHgYDmpraXzY0AMC//42LF3HlCpyckJmJZcuQkQEnpy40f+7cuebm5ldeeUWhUNioYqLuwBilzmhsRG6uZdtoBICzZ/H225bcnDgRKhWysjB4MNzd4ezcqbZv3749adKkPn362LZkIqExRqkziotx8KBlu60NAKqrMWhQ+w5ubqiuRkgIqqowdChGjcLIkRg1yrzROnq03MPjl9retm1bWlraxYsXX375ZUEPgsi2GKPUGf7+OHzYsp2YCACjRyMvD8HBljdzcqBSQaGAoyPKy1FejpSUx5/+9vnn56emDh8+fNT/Gjt2rIuLS1JSUkVFxfLly7v5mIisxBgl64SFYf58+Ptj7FgcPw6VCj4+KCiAwYDSUqjVKC6GWm3eyJPJdDqdWq1Wq9VPtjF27Nj8/PxXX31VrIMgsgZjlDps8GDMndv+0nzWftQofPUVDh/GgweYNg2nT1t+K5XCywteXk828C6wuqGhuLhY/SPztre3d3cdA5Ht8fJ7IiKr8LpREp/JZLp58+bWrVtPPx7MEj09GKMkvpCQkDVr1igUiilTpohdC1GncVJP4tNoNIOevGqK6KnC0SiJz5yh9fX1sbGx5eXlYpdD1DmMUbILsbGxnp6ecXFxdXV1YtdC1Dmc1JNdqK6udnR0HDBggNiFEHUaR6NkF9zc3MwZmp+fL3YtRJ3DGCV7UVVVNWHChLlz51ZXV4tdC1EncFJP9sJkMqWlpU2dOlUikYhdC1EnMEbJ7phMJiYpPUU4qSf78v777//ud78TuwqiTuBolOzL1atXvb29lUql4D21tcFkan9Qv9EIB4ef2Sb6LYxR6n1qarBuHRoaIJFALsexY3B3x/DhKCuz7LBgAXbtQkCAqFXSU4P/csnuXLp0af/+/QJ2sG0b5szBpUv4+mu8+iq2bBGwL+oF+LxRsjseHh7Crsh04QKioy3bq1fjgw8AwGjE9euWNzUaAXunHocxSnZnwoQJwnbQ3Ay53LItlcJkgtEIvR7nzlnefPBA2AKoZ+GknuxRdXV15o9L3ldVVQUFBVnb4s2bCAiAefEST08UFlreLyuDmxscHODkhD//2fIzbpy13VFvwhgle9SvX7/Kykrz9t///vfFixd3vS2TCR9+iEWL8P332LgRACIjERqKO3eQkYG1axEZaYuSqffipJ7skVwuf/HFFwHU1dVJpdKbN2/m5eX5+Ph0uqHKSqxahatX0dICANevIzkZf/gDlEqcOgWjEX/6E2bPBoCwsPZPLVoEd3fbHAn1ArzgiexRc3Pzhg0bCgoK1q5du3r16rKyshEjRnS6lSNHEBVlmcj37QsHB8hk2LoV27fbvGDqzRijZHcuX778zjvv5OXlSaVSg8EwZsyY8PDwN954QyaTdbSJR48wbx5yciCXw2SCXo/Bg7FtG9atg+P/zMAyMzO//fbbd99994svvjAYDKGhobY/Hurp+N0o2ZGysrLXXntt9uzZeXl5/v7+y5YtGzFiRGFhYVhY2NixYw8dOtTU1PSbjXz99df7/PyMd++ipQV9+yIwEOfP4949vP32TzIUwMSJE/V6/f3796VSqaura05OjjBHRj0ZY5Tsgl6vj46O9vPzS0hIUCgUO3fuvHXrVmxsbFFR0YkTJ/z8/O7du7dx40aVSrVr166ampqfbaStrW39+vUrVqyIfPQoS6HA6tW4fh3nz2PmzF/v3TzsNRgMUqlUgIOjns5EJLaUlBR/f3/zH+T8+fPv3bv3kx0MBkNiYuLvf/978z79+vXbtGlTaWnpk/sUFBQEBQU5OztPmjRp69atGo2mI12r1ep9+/YVFhbGxMR8+eWXtjoi6lUYoySqH34wvfnmvmefBTBmzJgLFy78+u4pKSnz5883P0bPyclp5cqV2dnZJpNp9+7dAwYMmD59ekxMjMFg6JbSiSx4iolEYjLhiy8QGYnqar1K9cmaNeGRkc7Ozh356O3bt/fs2ZOQkGA0GqVS6ejRo93d3fft2zdt2jShqyb6/xijJIbMTKxfj9RUAJg1C59/Dl/fTjVQUlKi1Wq3b9+emJgok8l0Op0gdRJ1AE8xkW0cOHAgMTHxxo0bkb9+U1BTE3btwtSpSE2FhwdOnMDly53N0LKysqKiotOnT//zn/90dHQ0GAyNjY1WVU9kBcYo2cbChQtNJtP06dM9PDza362tRUICYmIs97BfuQJvb3z0EQwGbN6M3FysWtWFvoYPH67X6wcNGuTs7Dx+/Hij0fjf//7XRsdB1GmMUbKNxsZGrVbb0tLS3NxsmWIXF+P551FSgrY2rF6Nf/wD7u54+BABAbh+HQcOwMWla32VlJTo9fr6+nr8+Dio7Oxs2x0KUefwnnqyjYaGBg8Pj6qqqilTptTW1iqVSuzejchIvP46AISEYOpUqNW4ehVTpli5RIeXl5eTk9O8efMAjB8/HsDdu3dtcRBEXcEYJdsIDAw0b3h5eVneSk/Hzp2W7SFD4OqKigrY6GT60KFDzRvmC06zsrJs0ixRF3BST4JxckJbW/tLna79Ycm2w0k9iY4xSoIJDMTFi5ZttRomE4YMsXknKpXKxcWlsrLy0aNHNm+cqCM4qSfBbNuGhQuRkwNXV/zrX/j8cyE6kUgkfn5+33333d27d4ODg4XogujXcTRKghk8GN98gzfewOzZuHgRzz8vUD+c15O4OBolITk6QqnE/PlwdrY8PlkAjFESF2OUBObujkePYDCgpQUdu2W+sxijJC5O6klgMhlGjoTR2L4Yp62ZYzQrK4sPiCBRMEZJeOal6PLyBGre3d19yJAhdXV15eXlAnVB9CsYoyQ8gWMUTwxIheuC6JcwRkl4wseo+V4mfj1KomCMkvCEj1HeWU8iYoyS8Dippx6NT7+n7qD39m5QKJwuXlQolUK039DQMGDAALlc3tjYyNU9qZtxNErdIcjV1TU9/Y5gA9L+/ft7enq2tLQUFRUJ1AXRL2GMUnfw8fEBkCdAjF64cCEqKio9PZ0X4ZNYGKPUHYSL0ZdeemnOnDkVFRWMURILY5S6g3Axmp+ff+fOnZdfflkmkwE4deqUzbsg+nWMUeoO48aNgzAxWlhYqFAodu/evX//fgBKpfLGjRslJSWor0ddHQDwJCoJjDFK3WHMmDFSqVStVtt2QXmDwVBVVRUZGbljxw6tVuvm5hYVFZWUlFRQUIAjRxARgYICTJ8OAOfPY9YsvPAC5szB9esAEBWF+HhLQykpeP99GxZGvQqf8ETdQS6Xq1QqtVqtVqvNI1PrJScnb9myJTMzE8D48eN37ty5ZMkSPF4V6sMPsXcvYmLw3HMoKkJ4OK5cwZAhKC7GvHn47jvU10OrtbTV0gKNxiZVUS/E0Sh1E/PXo9988431TV27dm3mzJlz587NzMxUqVRHjhzJyMgwZ2i75GTMmIHYWKxcicRELF9uWcJk5EjMm4fkZACorUVFBSoqUFNjfVXUa3E0St1kyJAhMplsy5YtLS0tGzdudOjSGstZWVkff/xxfHy8ucEtW7a899578p9dKa9fPxgMiI7GM88gLg4jR7b/SqnEw4cAEBeHW7cAoKoKI0Z0oR4icDRKQjt48OCePXsKCgr27t3r5eWl1Wrfe++9mTNnZmRkdKqdvLy8pUuXTpo0KT4+vn///hEREUVFRRERET+foWZSKebPBwCVCk9ell9QYEnV0FCcOoVTpxAR0ZVjIwLAGCWh1dXVhYaGXrp0SalU5ufnJyYmjhgx4tq1a1OmTNm8eXN9ff2TO8fFxR09evQnLZSVlYWFhU2YMCE+Pr5Pnz6bNm0qKiras2dP//79O1rEsmU4cwYpKdDpcO4cMjPx4os2OToiMEZJaK2trUVFRcOGDTO/DAkJycnJiYiIAHDw4MFx48adPHny8YMdFi1apNfrH3+2pqYmMjLS29v76NGjEokkNDS0oKAgOjp6SGcXah44EOfPIzYWISFITkZyMmQyTJgALy/LDu7umDrV2kOlXstEJKSioqIzZ84YjcafvJ+enj5jxgzzH2FwcHB2drbJZKqurt67d6/RaGxsbNyzZ8/AgQMBSCSSJUuW5Ofni1E+0W/jE55INCaTKSYmJjw8/OHDhzKZbP369S+99FJdXZ1Go/n4448rKysBzJkz59NPPw0ICBC7WKJfxBglkdXU1Gzbtu3YsWNGo3HQoEGtra1NTU0AAgMDP/nkk5kzZ9qwL41GExMT4+XltWDBAhs2S70cvxslkbm6uh45ciQtLW3atGn19fVNTU0qlSouLu7q1au2zVAA165de/bZZ1NSUmzbLPVyHI2SvWhra1u/fn1paWlSUpKjoyBXNBsMhjNnzvznP/85fPiwEO1T78TRKNkLmUy2YMECnU4nUIYC0Gq1Go1m5cqVArVPvRPvYiI74ujoqFarhWvfxcVl3bp1wrVPvRNHo2RHJBJJY2Njc3Oz2IUQdQJjlOyIo6NjY2Pj999/L3YhRJ3AGCU7IpVKdTrd+fPnxS6EqBMYo2RHzGsjc7l5erowRsmOODg4SCTSe/dqxS6EqBN4pp7sSEuLb58+P+TmKpqa0Lev2NUQdQxHo2RHFIrBOt0gwOn2bbFLIeowxijZEYkEEgmkUvAkEz1FGKNkRxwcYDLBaMTdu2KXQtRhjFGyIw4OkErR3IzKSrFLIeowxijZEfNoFMCDB2hsFLsaoo5hjJIdcXCAecHQqirLkp1E9o8xSnZEIoH56U46Hc8y0VODMUoia21tfbzd0ACjEc7OcHdHQYGIRRF1Ah/bTGLasWOHXC5fsWKFSqUC0NiIM2fg44Nhw+DuDqlU7PqIOoCjUeq6v/3tb6GhoeXl5V1uobW1tbi42MXFxfyyogK3b+OZZzB0KGpqcOAAAHzwQfv+0dH44QeraiayOcYodd1bb73l6+v7eA36LjCZTK+88srjZ5FUVuL4cXz2GQDU1yMpCQBOnGjf/+xZ1PKGe7IzjFHquszMTCuXPt68ebOXl9eTS9ctWoS4OJSUWFsbUbfho0mo6yQSSXBwsDUtDBs27CeDWScnfPIJNm60zOgBVFfjhRcs2+np1vRGJAiORqkrdDrdsmXLzAvK29y8eejTxzKjB+DmhsuXLT/PPCNEh0RWYYxSVxw+fDguLi4oKGjXrl16vd76BktKsH9/+8vPPrN8Q0pk/xij1BUbNmzYuXOnyWT66KOPAgMD8/LyrGntq68weTLefx9Xr1reGT4c775r2XZyat9TJoNEYk1XRLbH60ap665fv75q1arCwsI+ffrs3LkzPDzcwaFz/5hbWhARgYMHAWDhQhw/DldXQUolEg5jlKxSX18fHh5+9OhRAHPnzv3yyy87fv1Tbi5efx0ZGZDL8emn2LSJI016KjFGyQYSEhLCwsIePXo0cODAQ4cOLV++/Dc/cvIk3nkHWi18fHD6NM8d0VOM342SDSxevPju3bsLFiyora1dsWLF0qVLa2pqHv923759x48ff3wmqqGhYcWKFUePFmm1WLkSt24xQ+npxhgl21AqlWfPnj1x4kS/fv3i4+MnT5585coVAEaj8caNGxqNpqWlBcCtW7cCAgJiY2MLC1+MjdWdPIl+/cQuncg6jFGypVWrVmVmZgYFBd2/f3/27NlhYWEtLS0+Pj5KpfL+/fvR0dEzZswoLCwMCAhISTn/xz86/XaLRHaP342S7en1+r/+9a87duzQ6XR+fn5/+ctfZDLZ4cOHk5KSJBLJxo0bo6Ki5HK52GUS2QZjlISSlpa2ePHisrIyqVTq7Oys1WqlUmlcXNzixYvFLo3IljipJ6FMnTp10aJFAAwGg/kdpVLJDKWehzFKAsrPzwewffv2t956C4C/v7/YFRHZHmOUBJSbmwtgzZo1Q4cOBTBu3DixKyKyPcYoCUWr1ZaWljo5OY0cOdKcp4xR6pEYoySU3Nxck8nk7e3t6OiYk5MDwNfXV+yiiGyPMUpCMY9AzdH55DZRD8On35NQ/B88OBQYOCAg4MGDBxqNZuDAge7u7mIXRWR7jFESysTU1ImpqdiwwZST0zZs2A9BQWJXRCQITupJMDk5AODrK8nJcSwv91AoxC6ISBCMURKGXo+iIkgk8PZGbi4A8DQ99VCMURKGWo3WVnh6QqGwxCjPL1EPxRglYfw4o2/f5miUeijGKAnj8Qi0sRFlZZDL4eUlcklEwmCMkjDkcoweDT8/5ObCZIK3Nxx5WQj1THxQHglGp8PDh3B2xpUrALBkidgFEQmCo1ESxvbtCA7G9u2YPRtOTsxQ6sE4zyIBJCUhOxupqZBIUFODadMQFAQ3N7HLIhIER6MkgORkLFtmWXXe1RWBgbh1S+yaiITCGCUBaLXo27f9pUKBxkbxqiESFmOUBODri/T09pd37sDPT7xqiITF70ZJAGvX4rnn4OmJiRMRF4exY3kLE/VgvOCJhFFbi5MnUV6OyZOxdCkcOO+hHosxSkRkFY4RiIiswu9GycZaW1ujoqKWLl1aWlqal5cXFBQ0adIksYsiEhBHo2Rjcrk8ODi4qalp1qxZZWVljryVnno6xigJRaPRhIWF3eKF99TTcaRAtpebm9u/f3+TyZSRkfHaa6+JXQ6RsHimnojIKpzUExFZhTFKRGQVxigRkVUYo0REVmGMEhFZhTFKRGSV/wPr3jq7g2TBbgAAAVV6VFh0cmRraXRQS0wgcmRraXQgMjAyMy4wOS41AAB4nHu/b+09BiDgZ0AAPihuYGRjSADSTExsDClAmpGZEcJggjEYURgwtRpAmpkFwmdkZmfIAPGZGDkgAki0AkgDggvWx8TNwMjAACSYGBg5WBiZmBmYOFmYmFkYmLlYGFlYWZhYuBlYeRjY2BnYOBTYeDWYRIDKGdgYuFmAVohvAhkFxQx8IckCTuI3jzoWZ0xx1I8Vdiq0PeTol9TlOO2whNP578cd95R0OQbtEHcK47viKLKuy/FqhYjTGpnrjv2Hpji6dAs5LZv9yLH5wFTHg6f4na4v3uD45W+bI9d1bqeLStscJ0vUOe75IuzUf2ibo/LtTkfJFmmnbZ+POu6b3+BovFnGSXHDTUd70y5Hww5+pySBi45PL0x2LLTldeJheuI47/1MxwxBfie2xuWOs2Z2OYoBADHoU95HFAmgAAABoHpUWHRNT0wgcmRraXQgMjAyMy4wOS41AAB4nH1Ty24CMQy88xX5gii28/KhB1goqioWqaX9h977/6qdB9kg1N09xGbGGY8N2GVn9Pk4vv/8mvtDx53k3T8fM5tvcs7tLkYP5nA6v61mue0PPbNcv9bbpwGvn9N3xu5v10vPgFkMRes4SUmD3iYCPUG2IZTTnYwVmj00KGZXoMki5wI1y+v5BTqBlJAtcCcwx0YglwYBO8ErIVmmUhejDRWmBApPbgiVAJAbIWfu6vkZIdYeIhUhmC1hbIQEPLWbFBos+1obrWvtyi3JT9CsUGkvugZNseoOImg2kc2qArjqFGj2oZtIOEHBmav6lxP32XjfFLiUZywoli2kUg2lGkCvm+KMRcWqNGquAVEzwac0Y6liy1IUw4Krddmif9DrK7Y1j04Gmfv4MEzY03qc1rAu5uG6Hsdi6otj+UoCx27Jj4bG5mjIj9sBgvNj9hpC3QgaYwZhhzFKDcXOqRJLNujybJngSnZoApUot9DGaZAL8sZNjXnjGKoAv7Vla4LG/Z8t590fVI/L5SZwMYQAAAE3elRYdFNNSUxFUyByZGtpdCAyMDIzLjA5LjUAAHicTZCxagQxDER/JeUdGGFJtiV5CQQ2RarbNKlCihQpw4ZwcM19fCRDWFfWjEfSs9f19LidL+/r08sHrts4T+t29iKq/wK3h/uJC1hLRCBNElZQxYUrWNEwsyZsoFLCiwRlMHZTgKku3MDUIqilhklMYWrBRAVoBMk0vGwSnjAmVKh1rFHlRA2QOcwisrAAooapPjmSNvY0dkr1rS1MQYt2n+VezT7SgApFt42JNcuAZG82QKmJnARxAElbWAFtQJo/P5JZwlSxQVlKPDyLntPndf9+/d1/eoYoL/v1C6TjIVqnQyB3PpT2Ml31eojc23SDXQ5Vuh6idpty1HGi4I4TBuaOE4hHJxBLtzdn8W/OHrw9+z7/IHLc+x+PwYw6m/aWBQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<rdkit.Chem.rdchem.Mol at 0x1c9de45f0>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.resolved_ligand_mols[\"1.C\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ligand conformer coordinates\n",
    "\n",
    "As you can tell, the input 2D and 3D conformer indices match, but the resolved ligand is different.\n",
    "Thus to perform a correct comparison for their coordinates one should use atom order stacks.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_atom_order_stack, holo_atom_order_stack = holo_struc.ligand_template2resolved_atom_order_stacks[\"1.C\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[36.097, 24.731, 18.551],\n",
       "       [36.841, 24.28 , 17.298],\n",
       "       [38.191, 24.996, 17.307],\n",
       "       [37.93 , 26.507, 17.335],\n",
       "       [37.118, 26.889, 18.595],\n",
       "       [36.636, 28.326, 18.719],\n",
       "       [35.948, 22.08 , 16.874],\n",
       "       [34.96 , 22.767, 15.881],\n",
       "       [36.989, 22.845, 17.232],\n",
       "       [38.879, 24.744, 16.078],\n",
       "       [39.175, 27.211, 17.276],\n",
       "       [35.883, 26.133, 18.477],\n",
       "       [35.31 , 28.501, 19.242],\n",
       "       [35.767, 20.938, 17.325]])"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.resolved_ligand_mols_coords['1.C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[[-4.16160949,  0.79077189, -0.70202375],\n",
       "        [-2.6793468 ,  0.73697006, -0.95195015],\n",
       "        [-2.16601153,  1.18459397, -1.97369102],\n",
       "        [-1.95094258,  0.14911666,  0.05385442],\n",
       "        [-0.48018606,  0.15936743,  0.04919472],\n",
       "        [ 0.12797791, -0.55975376, -1.17665183],\n",
       "        [ 0.39265184,  0.3993327 , -2.21036927],\n",
       "        [ 1.42867754, -1.30354759, -0.83915609],\n",
       "        [ 1.0832802 , -2.61950319, -0.37342512],\n",
       "        [ 2.24431263, -0.58722499,  0.2514995 ],\n",
       "        [ 2.89791411,  0.70636251, -0.25439678],\n",
       "        [ 3.77739399,  1.21614722,  0.75129176],\n",
       "        [ 1.4997948 , -0.25733533,  1.44622   ],\n",
       "        [ 0.09412268, -0.43743745,  1.3481003 ]]])"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holo_struc.input_ligand_conformers_coords['1.C'][holo_atom_order_stack]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Interacting with PLINDER systems\n",
    "PlinderSystem is the next layer of abstraction above `Structure`. It provides encapsulation around all structures associated with a particular `system_id`. With these, we can access the `holo` and alternate (`apo` and `pred`) structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_system = PlinderSystem(\n",
    "    system_id=\"1avd__1__1.A__1.C\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check holo structure\n",
    "Since having `holo` structure is a defining feature of _PLINDER_ system, holo structures is by definition available for all systems "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-23 12:31:02,302 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 12:31:02,302 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 12:31:03,078 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 12:31:03,080 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 12:31:03,081 | plinder.core.index.utils:148 | INFO : loading entries from 1 zips\n",
      "2024-09-23 12:31:03,087 | plinder.core.index.utils:163 | INFO : loaded 1 entries\n",
      "2024-09-23 12:31:03,088 | plinder.core.index.utils.load_entries:24 | INFO : runtime succeeded: 0.78s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Structure(\n",
       "    (\n",
       "        'id',\n",
       "        '1avd__1__1.A__1.C',\n",
       "    ),\n",
       "    (\n",
       "        'protein_path',\n",
       "        /Users/vladas/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/receptor.cif,\n",
       "    ),\n",
       "    (\n",
       "        'protein_sequence',\n",
       "        {\n",
       "            '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'ligand_sdfs',\n",
       "        {\n",
       "            '1.C': '/Users/vladas/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/ligand_files/1.C.sdf',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'ligand_smiles',\n",
       "        {\n",
       "            '1.C': 'CC(=O)N[C@@H]1[C@@H](O)[C@H](O)[C@@H](CO)O[C@H]1O',\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'protein_atom_array',\n",
       "        <class 'biotite.structure.AtomArray'> with shape (964,),\n",
       "    ),\n",
       "    (\n",
       "        'ligand_mols',\n",
       "        {\n",
       "            '1.C': (\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x1d066b760>,\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x1d0d05770>,\n",
       "                <rdkit.Chem.rdchem.Mol object at 0x1d0b71690>,\n",
       "                (\n",
       "                    <class 'numpy.ndarray'> with shape (1, 14),\n",
       "                    <class 'numpy.ndarray'> with shape (1, 14),\n",
       "                ),\n",
       "            ),\n",
       "        },\n",
       "    ),\n",
       "    (\n",
       "        'add_ligand_hydrogens',\n",
       "        False,\n",
       "    ),\n",
       "    (\n",
       "        'structure_type',\n",
       "        'holo',\n",
       "    ),\n",
       ")"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.holo_structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get annotations \n",
    "This `system` property returns `json` data of annotations for a the system in question. To get the annotations of all other systems sharing the same PDB entry ids, use `.entry` property."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pdb_id': '1avd',\n",
       " 'biounit_id': '1',\n",
       " 'ligands': [{'pdb_id': '1avd',\n",
       "   'biounit_id': '1',\n",
       "   'asym_id': 'C',\n",
       "   'instance': 1,\n",
       "   'ccd_code': 'NAG',\n",
       "   'plip_type': 'SMALLMOLECULE',\n",
       "   'bird_id': '',\n",
       "   'centroid': [36.85636520385742, 25.090288162231445, 17.591215133666992],\n",
       "   'smiles': 'CC(=O)N[C@@H]1[C@@H](O)[C@H](O)[C@@H](CO)O[C@H]1O',\n",
       "   'resolved_smiles': 'OC[C@H]1O[CH][C@@H]([C@H]([C@@H]1O)O)NC(=O)C',\n",
       "   'residue_numbers': [1],\n",
       "   'rdkit_canonical_smiles': 'CC(=O)N[C@@H]1[C@@H](O)[C@H](O)[C@@H](CO)O[C@H]1O',\n",
       "   'molecular_weight': 221.0899372,\n",
       "   'crippen_clogp': -3.077599999999999,\n",
       "   'num_rot_bonds': 2,\n",
       "   'num_hbd': 5,\n",
       "   'num_hba': 6,\n",
       "   'num_rings': 1,\n",
       "   'num_heavy_atoms': 15,\n",
       "   'is_covalent': True,\n",
       "   'covalent_linkages': ['17:ASN:A:17:ND2__600:NAG:C:.:C1'],\n",
       "   'neighboring_residues': {'1.A': [9, 11, 15, 16, 17, 34, 35, 36, 123]},\n",
       "   'neighboring_ligands': [],\n",
       "   'interacting_residues': {'1.A': [34, 15]},\n",
       "   'interacting_ligands': [],\n",
       "   'interactions': {'1.A': {'15': ['type:hydrogen_bonds__protisdon:False__sidechain:False'],\n",
       "     '34': ['type:hydrophobic_contacts']}},\n",
       "   'neighboring_residue_threshold': 6.0,\n",
       "   'neighboring_ligand_threshold': 4.0,\n",
       "   'num_neighboring_ppi_atoms_within_4A_of_gap': 0,\n",
       "   'num_neighboring_ppi_atoms_within_8A_of_gap': 0,\n",
       "   'num_missing_ppi_interface_residues': 0,\n",
       "   'num_pli_atoms_within_4A_of_gap': 0,\n",
       "   'num_pli_atoms_within_8A_of_gap': 0,\n",
       "   'num_missing_pli_interface_residues': 0,\n",
       "   'num_resolved_heavy_atoms': 14,\n",
       "   'num_unresolved_heavy_atoms': 1,\n",
       "   'tpsa': 119.25000000000001,\n",
       "   'qed': 0.33723889506423077,\n",
       "   'is_ion': False,\n",
       "   'is_lipinski': True,\n",
       "   'is_fragment': False,\n",
       "   'is_oligo': False,\n",
       "   'is_cofactor': False,\n",
       "   'in_artifact_list': False,\n",
       "   'is_artifact': False,\n",
       "   'is_other': False,\n",
       "   'is_invalid': False,\n",
       "   'posebusters_result': {'mol_pred_loaded': True,\n",
       "    'mol_true_loaded': True,\n",
       "    'mol_cond_loaded': True,\n",
       "    'sanitization': True,\n",
       "    'all_atoms_connected': True,\n",
       "    'molecular_formula': True,\n",
       "    'molecular_bonds': True,\n",
       "    'double_bond_stereochemistry': True,\n",
       "    'tetrahedral_chirality': True,\n",
       "    'bond_lengths': True,\n",
       "    'bond_angles': True,\n",
       "    'internal_steric_clash': True,\n",
       "    'aromatic_ring_flatness': True,\n",
       "    'double_bond_flatness': True,\n",
       "    'internal_energy': True,\n",
       "    'protein-ligand_maximum_distance': True,\n",
       "    'minimum_distance_to_organic_cofactors': True,\n",
       "    'minimum_distance_to_inorganic_cofactors': True,\n",
       "    'minimum_distance_to_waters': True,\n",
       "    'volume_overlap_with_protein': True,\n",
       "    'volume_overlap_with_organic_cofactors': True,\n",
       "    'volume_overlap_with_inorganic_cofactors': True,\n",
       "    'volume_overlap_with_waters': True,\n",
       "    'rmsd__2': True,\n",
       "    'passes_valence_checks': True,\n",
       "    'passes_kekulization': True,\n",
       "    'inchi_crystal_valid': True,\n",
       "    'inchi_docked_valid': True,\n",
       "    'inchi_crystal': 'InChI=1S/C8H15NO5/c1-4(11)9-5-3-14-6(2-10)8(13)7(5)12/h5-8,10,12-13H,2-3H2,1H3,(H,9,11)/t5-,6+,7+,8+/m0/s1',\n",
       "    'inchi_docked': 'InChI=1S/C8H15NO5/c1-4(11)9-5-3-14-6(2-10)8(13)7(5)12/h5-8,10,12-13H,2-3H2,1H3,(H,9,11)/t5-,6+,7+,8+/m0/s1',\n",
       "    'inchi_overall': True,\n",
       "    'inchi_version': True,\n",
       "    'stereochemistry_preserved': True,\n",
       "    'hydrogens': True,\n",
       "    'net_charge': True,\n",
       "    'protons': True,\n",
       "    'stereo_sp3': True,\n",
       "    'stereo_sp3_inverted': True,\n",
       "    'stereo_type': True,\n",
       "    'number_bonds': 14,\n",
       "    'shortest_bond_relative_length': 0.9492650185663352,\n",
       "    'longest_bond_relative_length': 1.0429011482666541,\n",
       "    'number_angles': 19,\n",
       "    'most_extreme_relative_angle': 1.0452636472719188,\n",
       "    'number_noncov_pairs': 58,\n",
       "    'shortest_noncovalent_relative_distance': 0.7386533194416193,\n",
       "    'number_valid_bonds': 14,\n",
       "    'number_valid_angles': 19,\n",
       "    'number_valid_noncov_pairs': 58,\n",
       "    'aromatic_ring_maximum_distance_from_plane': None,\n",
       "    'double_bond_maximum_distance_from_plane': None,\n",
       "    'ensemble_avg_energy': 32.10719429562964,\n",
       "    'mol_pred_energy': 81.78525064878713,\n",
       "    'energy_ratio': 2.54725622848707,\n",
       "    'smallest_distance_protein': 1.4676218859093153,\n",
       "    'num_pairwise_clashes_protein': 2,\n",
       "    'most_extreme_protein_atom_id_protein': 111,\n",
       "    'most_extreme_ligand_element_protein': 'C',\n",
       "    'most_extreme_protein_element_protein': 'N',\n",
       "    'most_extreme_ligand_vdw_protein': 1.7,\n",
       "    'most_extreme_protein_vdw_protein': 1.6,\n",
       "    'most_extreme_sum_radii_protein': 3.3,\n",
       "    'most_extreme_distance_protein': 1.4676218859093153,\n",
       "    'most_extreme_sum_radii_scaled_protein': 3.3,\n",
       "    'most_extreme_relative_distance_protein': 0.44473390482100467,\n",
       "    'most_extreme_clash_protein': True,\n",
       "    'smallest_distance_organic_cofactors': None,\n",
       "    'smallest_distance_inorganic_cofactors': None,\n",
       "    'smallest_distance_waters': None,\n",
       "    'volume_overlap_protein': 0.052308591466978374,\n",
       "    'volume_overlap_organic_cofactors': None,\n",
       "    'volume_overlap_inorganic_cofactors': None,\n",
       "    'volume_overlap_waters': None},\n",
       "   'unique_ccd_code': 'HSR',\n",
       "   'waters': {},\n",
       "   'crystal_contacts': {'B,125': [1946, 1947]}}],\n",
       " 'ligand_validation': {'num_residues': 1,\n",
       "  'num_processed_residues': 1,\n",
       "  'percent_processed_residues': 100.0,\n",
       "  'average_rsr': 0.192,\n",
       "  'average_rsrz': nan,\n",
       "  'average_rscc': 0.851,\n",
       "  'average_occupancy': 0.8571428571428571,\n",
       "  'percent_rsr_under_threshold': 100.0,\n",
       "  'percent_rscc_over_threshold': 100.0,\n",
       "  'percent_occupancy_over_threshold': 0.0,\n",
       "  'average_b_factor': 70.71357142857143,\n",
       "  'unknown_residue_count': 0,\n",
       "  'atom_count': 14,\n",
       "  'heavy_atom_count': 14,\n",
       "  'num_unresolved_heavy_atoms': 1,\n",
       "  'max_alt_count': 1,\n",
       "  'percent_outliers': {'geometry': 0.0,\n",
       "   'density': 0.0,\n",
       "   'chirality': 0.0,\n",
       "   'clashes': 0.0}},\n",
       " 'pocket_validation': {'num_residues': 9,\n",
       "  'num_processed_residues': 9,\n",
       "  'percent_processed_residues': 100.0,\n",
       "  'average_rsr': 0.12877777777777777,\n",
       "  'average_rsrz': -0.6723333333333333,\n",
       "  'average_rscc': 0.9612222222222222,\n",
       "  'average_occupancy': 0.9367283950617283,\n",
       "  'percent_rsr_under_threshold': 100.0,\n",
       "  'percent_rscc_over_threshold': 100.0,\n",
       "  'percent_occupancy_over_threshold': 77.77777777777777,\n",
       "  'average_b_factor': 31.81567680776014,\n",
       "  'unknown_residue_count': 0,\n",
       "  'atom_count': 61,\n",
       "  'heavy_atom_count': 61,\n",
       "  'num_unresolved_heavy_atoms': 0,\n",
       "  'max_alt_count': 1,\n",
       "  'percent_outliers': {'geometry': 0.0,\n",
       "   'density': 0.0,\n",
       "   'chirality': 0.0,\n",
       "   'clashes': 0.0}},\n",
       " 'pass_criteria': False}"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.system"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get paths of the underlying structure files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`archive` points to the subfolder where all the files (except `apo` and `pred` files) relating to a given system are stored"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/vladas/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C')"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.archive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similarly, the  `system.cif`, `receptor.cif`, `receptor.pdb`, `sequence.fasta`, and ligand sdfs can be accessed via `.system_cif`, `.receptor_cif`, `.receptor_pdb`, `.sequence_fasta` and `.ligand_sdf` properties respectively. \n",
    "\n",
    "To get all the paths of the structures together, use `.structures`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/Users/vladas/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/receptor.cif',\n",
       " '/Users/vladas/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/receptor.pdb',\n",
       " '/Users/vladas/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/chain_mapping.json',\n",
       " '/Users/vladas/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/sequences.fasta',\n",
       " '/Users/vladas/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/system.cif',\n",
       " '/Users/vladas/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/ligand_files/1.C.sdf']"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/vladas/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/system.cif'"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.system_cif"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get binding site water (`.water_mapping`)\n",
    "This returns the information about binding site water "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_system.water_mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Chain mapping \n",
    ":::{todo}\n",
    "Confirm with Jay\n",
    ":::\n",
    "\n",
    "`.chain_mapping` maps chain ids in system (`<instance_id>.<asym_id>`) to PDB author chain ids "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1.A': 'A'}"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.chain_mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "linked_structures, linked_archive, get_linked_structure \n",
    "linked_archive, linked_structures,  get_linked_structure, alt_structures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Linked apo and predicted structures.\n",
    "The following properties provides different kind of information about linked structures as described below:\n",
    "- `.linked_archive`: returns paths the local subfolder where the linked structures are saved; \n",
    "- `.linked_structures`: returns the dataframe of linked structures along with all their metrics while \n",
    "- `.get_linked_structure`:  gives the path to a specific linked structure.\n",
    "- `.best_linked_structures_paths`: Gives the best linked structures based on `scrmsd_wave` which is average symmetry-corrected RMSD across mapped ligands weighted by number of atoms. This selects maximum of two alternate structure with at most one `apo` and `pred` each when available\n",
    "- `.alt_structures`: returns the dictionary`Structure` object of the best `apo` and `pred` which the corresponding `holo` chain as key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-23 12:02:08,163 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-23 12:02:08,164 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c73ece447b8c40528feea694c120cab5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/11 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/vladas/.local/share/plinder/2024-06/v2/linked_structures')"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_system.linked_archive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'PlinderSystem' object has no attribute 'linked_structure'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[80], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43msample_system\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlinked_structure\u001b[49m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m# TODO AttributeError: 'PlinderSystem' object has no attribute 'linked_structure'\u001b[39;00m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'PlinderSystem' object has no attribute 'linked_structure'"
     ]
    }
   ],
   "source": [
    "sample_system.linked_structure\n",
    "# TODO AttributeError: 'PlinderSystem' object has no attribute 'linked_structure'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/1vyo_B.cif'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mnotebook controller is DISPOSED. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "sample_system.get_linked_structure(link_kind=\"apo\", link_id='1vyo_B')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1vyo_B': Structure(\n",
       "     (\n",
       "         'id',\n",
       "         '1vyo_B',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/1vyo_B.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (958,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'apo',\n",
       "     ),\n",
       " ),\n",
       " '1vyo_A': Structure(\n",
       "     (\n",
       "         'id',\n",
       "         '1vyo_A',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/1vyo_A.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (947,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'apo',\n",
       "     ),\n",
       " ),\n",
       " '1nqn_A': Structure(\n",
       "     (\n",
       "         'id',\n",
       "         '1nqn_A',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/1nqn_A.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (928,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'apo',\n",
       "     ),\n",
       " ),\n",
       " '1rav_B': Structure(\n",
       "     (\n",
       "         'id',\n",
       "         '1rav_B',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/1rav_B.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (976,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'apo',\n",
       "     ),\n",
       " ),\n",
       " '1rav_A': Structure(\n",
       "     (\n",
       "         'id',\n",
       "         '1rav_A',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/1rav_A.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (976,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'apo',\n",
       "     ),\n",
       " ),\n",
       " 'P02701_A': Structure(\n",
       "     (\n",
       "         'id',\n",
       "         'P02701_A',\n",
       "     ),\n",
       "     (\n",
       "         'protein_path',\n",
       "         /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/P02701_A.cif,\n",
       "     ),\n",
       "     (\n",
       "         'protein_sequence',\n",
       "         {\n",
       "             '1.A': 'ARKCSLTGKWTNDLGSNMTIGAVNSRGEFTGTYTTAVTATSNEIKESPLHGTENTINKRTQPTFGFTVNWKFSESTTVFTGQCFIDRNGKEVLKTMWLLRSSVNDIGDDWKATRVGINIFTRLRTQKE',\n",
       "         },\n",
       "     ),\n",
       "     (\n",
       "         'ligand_sdfs',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'ligand_smiles',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'protein_atom_array',\n",
       "         <class 'biotite.structure.AtomArray'> with shape (1179,),\n",
       "     ),\n",
       "     (\n",
       "         'ligand_mols',\n",
       "         None,\n",
       "     ),\n",
       "     (\n",
       "         'add_ligand_hydrogens',\n",
       "         False,\n",
       "     ),\n",
       "     (\n",
       "         'structure_type',\n",
       "         'pred',\n",
       "     ),\n",
       " )}"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mnotebook controller is DISPOSED. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "sample_system.best_linked_structures_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'apo': {'1.A': Structure(\n",
       "      (\n",
       "          'id',\n",
       "          'linked_structures',\n",
       "      ),\n",
       "      (\n",
       "          'protein_path',\n",
       "          /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/1vyo_A.cif,\n",
       "      ),\n",
       "      (\n",
       "          'protein_sequence',\n",
       "          /Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/sequences.fasta,\n",
       "      ),\n",
       "      (\n",
       "          'list_ligand_sdf_and_input_smiles',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'protein_atom_array',\n",
       "          <class 'biotite.structure.AtomArray'> with shape (947,),\n",
       "      ),\n",
       "      (\n",
       "          'ligand_mols',\n",
       "          {\n",
       "  \n",
       "          },\n",
       "      ),\n",
       "      (\n",
       "          'add_ligand_hydrogens',\n",
       "          False,\n",
       "      ),\n",
       "      (\n",
       "          'structure_type',\n",
       "          'apo',\n",
       "      ),\n",
       "  )},\n",
       " 'pred': {'1.A': Structure(\n",
       "      (\n",
       "          'id',\n",
       "          'linked_structures',\n",
       "      ),\n",
       "      (\n",
       "          'protein_path',\n",
       "          /Users/yusuf/.local/share/plinder/2024-06/v2/linked_structures/P02701_A.cif,\n",
       "      ),\n",
       "      (\n",
       "          'protein_sequence',\n",
       "          /Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/sequences.fasta,\n",
       "      ),\n",
       "      (\n",
       "          'list_ligand_sdf_and_input_smiles',\n",
       "          None,\n",
       "      ),\n",
       "      (\n",
       "          'protein_atom_array',\n",
       "          <class 'biotite.structure.AtomArray'> with shape (1179,),\n",
       "      ),\n",
       "      (\n",
       "          'ligand_mols',\n",
       "          {\n",
       "  \n",
       "          },\n",
       "      ),\n",
       "      (\n",
       "          'add_ligand_hydrogens',\n",
       "          False,\n",
       "      ),\n",
       "      (\n",
       "          'structure_type',\n",
       "          'pred',\n",
       "      ),\n",
       "  )}}"
      ]
     },
     "execution_count": 301,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mnotebook controller is DISPOSED. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "sample_system.alt_structures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get `Openstructure` entities and views\n",
    "`.receptor_entity` returns receptor `mol.EntityHandle` object\n",
    "`.ligand_views` returns `mol.ResidueView` for all ligands\n",
    "\n",
    ":::{note}\n",
    "You must have Openstructure installed to use this property\n",
    ":::\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Others properties\n",
    "This includes:\n",
    "- `num_ligands`: Number of ligand chains\n",
    "- `smiles`: Ligand smiles dictionary\n",
    "- `num_proteins`: Number of protein chains\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Interacting with the PLINDER dataset\n",
    "`PlinderDataset` provides an interface to interact with _PLINDER_ data as a dataset. It is a subclass of `torch.utils.data.Dataset`, as such subclassing it and extending should be familiar to most users. Flexibility and general applicability is our top concern when designing this interface and `PlinderDataset` allows users to not only define their own split but to also bring their own featurizer.\n",
    "It can be initialized with the following parameters\n",
    "```\n",
    "Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame | None\n",
    "        the split to use\n",
    "    split : str\n",
    "        the split to sample from\n",
    "    split_parquet_path : str | Path, default=None\n",
    "        split parquet file\n",
    "    input_structure_priority : str, default=\"apo\"\n",
    "        Which alternate structure to proritize\n",
    "    featurizer: Callable[\n",
    "            [Structure, int], dict[str, torch.Tensor]\n",
    "    ] = structure_featurizer,\n",
    "        Transformation to turn structure to input tensors\n",
    "    padding_value : int\n",
    "        Value for padding uneven array\n",
    "    **kwargs : Any\n",
    "        Any other keyword args\n",
    "``` "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "from plinder.core.loader import PlinderDataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Make _PLINDER_ training dataset with default parameters\n",
    "When no parameter is set, PlinderDataset automatically defaults to the training set of the most current version of the dataset. To change this behaviour, we can explicitly pass a data frame to the parameter `df` or split file to `split_parquet_path`. Either of these must have at least two columns named `system_id` and `split`. This also use our default featurizer `plinder.core.loader.featurizer.featurizer`. \n",
    "NOTE: We have provided this `plinder.core.loader.featurizer.featurizer` as an example featurizer; users are encourage to use featurizers that suit their need."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-22 19:39:55,575 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.67s\n",
      "2024-09-22 19:39:57,535 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.69s\n",
      "2024-09-22 19:39:58,208 | plinder.core.split.utils:40 | INFO : reading /Users/vladas/.local/share/plinder/2024-06/v2/splits/split.parquet\n",
      "2024-09-22 19:40:03,154 | plinder.core.split.utils.get_split:24 | INFO : runtime succeeded: 6.46s\n"
     ]
    }
   ],
   "source": [
    "train_dataset = PlinderDataset(split=\"train\")\n",
    "#train_dataset = PlinderDataset(df=splits_df[splits_df.system_id ==\"6pl9__1__1.A__1.C\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-22 19:40:16,271 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 8.52s\n",
      "2024-09-22 19:40:16,925 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.65s\n",
      "2024-09-22 19:40:22,063 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 2.90s\n",
      "2024-09-22 19:40:22,740 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.68s\n",
      "2024-09-22 19:40:22,742 | plinder.core.index.utils:148 | INFO : loading entries from 1 zips\n",
      "2024-09-22 19:40:22,749 | plinder.core.index.utils:163 | INFO : loaded 1 entries\n",
      "2024-09-22 19:40:22,750 | plinder.core.index.utils.load_entries:24 | INFO : runtime succeeded: 4.24s\n"
     ]
    }
   ],
   "source": [
    "test_data = train_dataset[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "test_data[110]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Loader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mnotebook controller is DISPOSED. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "train_loader = get_torch_loader(\n",
    "    train_dataset\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for data in train_loader:\n",
    "\n",
    "    test_torch = data\n",
    "    break\n",
    "    #for k, v in test_torch['input_features'].items():\n",
    "    #    if v.shape[1] > 1:\n",
    "    #        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_torch.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_torch['system_ids']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k, v in test_torch['features_and_coords'].items():\n",
    "    print(k, v.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "holo_struc.ligand_mols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "holo_struc.input_ligand_conformer2resolved_stacks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-20 16:58:27,992 | plinder.core.structure.atoms:158 | INFO : generate_conformer: MMFFOptimizeMolecule - more iterations are required, extending by 500 steps\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(PosixPath('/Users/yusuf/.local/share/plinder/2024-06/v2/systems/1avd__1__1.A__1.C/ligand_files/1.C.sdf'), 'CC(=O)N[C@@H]1[C@H]([C@@H]([C@H](O[C@H]1O)CO)O)O')]\n"
     ]
    }
   ],
   "source": [
    "holo_struc = Structure.load_structure(\n",
    "    id=system_id,\n",
    "    protein_path=protein_structure_path,\n",
    "    protein_sequence=input_sequence_path,\n",
    "    list_ligand_sdf_and_input_smiles=list_ligand_sdf_and_input_smiles\n",
    "\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-20 16:58:55,342 | plinder.core.structure.atoms:170 | WARNING : generate_conformer: EmbedMolecule - failed, try using useBasicKnowledge=False\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(PosixPath('/Users/yusuf/.local/share/plinder/2024-06/v2/systems/102m__1__1.A__1.C/ligand_files/1.C.sdf'), 'C=CC1=C(C)C2=Cc3c(C)c(CCC(=O)O)c4n3[Fe]35<-N6=C(C=c7c(C=C)c(C)c(n73)=CC1=N->52)C(C)=C(CCC(=O)O)C6=C4')]\n"
     ]
    }
   ],
   "source": [
    "struct = Structure.load_structure(id=\"102m__1__1.A__1.C\",\n",
    "          protein_path=Path(\"/Users/yusuf/.local/share/plinder/2024-06/v2/systems/102m__1__1.A__1.C/receptor.cif\"),\n",
    "          list_ligand_sdf_and_input_smiles=[(Path(\"/Users/yusuf/.local/share/plinder/2024-06/v2/systems/102m__1__1.A__1.C/ligand_files/1.C.sdf\"),\n",
    "          \"C=CC1=C(C)C2=Cc3c(C)c(CCC(=O)O)c4n3[Fe]35<-N6=C(C=c7c(C=C)c(C)c(n73)=CC1=N->52)C(C)=C(CCC(=O)O)C6=C4\")],\n",
    "          protein_sequence=Path(\"/Users/yusuf/.local/share/plinder/2024-06/v2/systems/102m__1__1.A__1.C/sequence.fasta\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-22 23:49:15,366 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-22 23:49:15,367 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-22 23:49:15,505 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-22 23:49:15,505 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-22 23:49:15,505 | plinder.core.index.utils:148 | INFO : loading entries from 1 zips\n",
      "2024-09-22 23:49:15,509 | plinder.core.index.utils:163 | INFO : loaded 1 entries\n",
      "2024-09-22 23:49:15,509 | plinder.core.index.utils.load_entries:24 | INFO : runtime succeeded: 0.14s\n",
      "2024-09-22 23:49:15,978 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.20s\n",
      "2024-09-22 23:49:16,176 | plinder.core.scores.links.query_links:24 | INFO : runtime succeeded: 0.56s\n",
      "2024-09-22 23:49:16,595 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n",
      "2024-09-22 23:49:16,595 | plinder.core.utils.cpl.download_paths:24 | INFO : runtime succeeded: 0.00s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'1.C': <rdkit.Chem.rdchem.Mol at 0x1d23d8120>}"
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "struct.input_ligand_conformers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [],
   "source": [
    "struct.ligand_mols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
